{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## effect size sharing for multi-gene qtls\n",
    "\n",
    "say we have a qtl that impacts multiple genes. How does the fractional effect of that qtl on each gene vary? w.r.t for instance distance from qtl to tss? maybe tie in ABC connections here? Maybe build a model that predicts how qtl impact on each gene will be split based on gene baseline expression, distance, ect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import ast\n",
    "import re\n",
    "\n",
    "import upsetplot as up\n",
    "from tqdm.auto import tqdm  # for notebooks\n",
    "\n",
    "# Create new `pandas` methods which use `tqdm` progress\n",
    "# (can use tqdm_gui, optional kwargs, etc.)\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get outputs from a config file\n",
    "prefix = '/home/klawren/oak/pcqtls'\n",
    "import yaml\n",
    "config_path= f'{prefix}/config/new_clusters/proteincoding_rewrite.yaml'\n",
    "with open(config_path, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "tissue_id_path = config['tissue_id_path']\n",
    "clusters_dir = config['clusters_dir']\n",
    "eqtl_output_dir = config['eqtl_output_dir']\n",
    "pcqtl_output_dir = config['pcqtl_output_dir']\n",
    "overlap_output_dir = config['overlap_output_dir']\n",
    "\n",
    "\n",
    "# load in the tissue ids \n",
    "tissue_df = pd.read_csv(f\"{prefix}/{tissue_id_path}\", header=0)\n",
    "tissue_ids = list(tissue_df['Tissue'])\n",
    "\n",
    "# select just 1 tissue id to do for now \n",
    "tissue_id = 'Cells_Cultured_fibroblasts' # because only one cell type\n",
    "tissue_ids = [tissue_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load in data\n",
    "effect sizes for qtls\n",
    "overlap dfs\n",
    "gencode for tss starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "513ce854297740698491bfbdada35f00",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/22 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ed20b43788464919b64ac18986894f9a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/22 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load nominal p value dfs\n",
    "\n",
    "# load in e nominal\n",
    "def load_e_nominal(path):\n",
    "    e_nominal_df = pd.read_parquet(path)\n",
    "    e_nominal_df['cluster_id'] = e_nominal_df['phenotype_id'].str.split('_e_').str[0]\n",
    "    e_nominal_df['egene_id'] = e_nominal_df['phenotype_id'].str.split('_e_').str[1]\n",
    "    return e_nominal_df\n",
    "\n",
    "e_nominal_dfs=[]\n",
    "for chr_id in tqdm(range(1,23)):\n",
    "    e_nominal_dfs.append(load_e_nominal(f'{prefix}/{eqtl_output_dir}/{tissue_id}/{tissue_id}.v8.cluster_genes.cis_qtl_pairs.chr{chr_id}.parquet'))\n",
    "\n",
    "e_nominal_df = pd.concat(e_nominal_dfs)\n",
    "\n",
    "def load_pc_nominal(path):\n",
    "    pc_nominal_df = pd.read_parquet(path)\n",
    "    pc_nominal_df['cluster_id'] = pc_nominal_df['phenotype_id'].str.split('_pc').str[0]\n",
    "    pc_nominal_df['pc_num'] = pc_nominal_df['phenotype_id'].str.split('_pc').str[1]\n",
    "    return pc_nominal_df\n",
    "\n",
    "pc_nominal_dfs=[]\n",
    "for chr_id in tqdm(range(1,23)):\n",
    "    pc_nominal_dfs.append(load_pc_nominal(f'{prefix}/{pcqtl_output_dir}/{tissue_id}/{tissue_id}.v8.pcs.cis_qtl_pairs.chr{chr_id}.parquet'))\n",
    "pc_nominal_df = pd.concat(pc_nominal_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_df = pd.read_csv(f'{prefix}/{overlap_output_dir}/{tissue_id}.v8.overlap.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# label cluster-variant pairs\n",
    "overlap_df['var_cluster'] = overlap_df['lead_variant_id'] + '_' + overlap_df['cluster_id']\n",
    "pc_nominal_df['var_cluster'] = pc_nominal_df['variant_id'] + '_' + pc_nominal_df['cluster_id']\n",
    "e_nominal_df['var_cluster'] = e_nominal_df['variant_id'] + '_' + e_nominal_df['cluster_id']\n",
    "\n",
    "# set e_nominal to be index based on variant-cluster pairs\n",
    "e_nominal_df_cid = e_nominal_df.set_index('var_cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in the susie df, so I have the pip values for the variants as well as the slopes\n",
    "pc_susie_df = pd.read_csv(f'{prefix}/{pcqtl_output_dir}/{tissue_id}/{tissue_id}.v8.pcs.susie.txt', sep='\\t', index_col=0)\n",
    "pc_susie_df['var_cluster'] = pc_susie_df['variant_id'] + '_' + pc_susie_df['phenotype_id'].str.split('_pc1').str[0]\n",
    "pc_susie_df['cs_id'] =  pc_susie_df['phenotype_id'].str.split('_pc1').str[0] + '_' + pc_susie_df['cs_id'].astype(str)\n",
    "pc_susie_df['pc_num'] = pc_susie_df['phenotype_id'].str.split('_pc').str[-1].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/klawren/slrmtmp.43212569/ipykernel_29162/3898079442.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  e_nominal_pcsusie_subset_df['var_egene_cluster'] = e_nominal_pcsusie_subset_df['variant_id'] + '_' + e_nominal_pcsusie_subset_df['phenotype_id']\n"
     ]
    }
   ],
   "source": [
    "# get a subset of variants that ended up in credible sets\n",
    "e_nominal_pcsusie_subset_df = e_nominal_df[e_nominal_df['var_cluster'].isin(pc_susie_df['var_cluster'])]\n",
    "\n",
    "# I want to get a pip weighted variance for each egene\n",
    "e_nominal_pcsusie_subset_df['var_egene_cluster'] = e_nominal_pcsusie_subset_df['variant_id'] + '_' + e_nominal_pcsusie_subset_df['phenotype_id']\n",
    "\n",
    "# merge in the data from the susie finemapping\n",
    "e_nominal_pcsusie_subset_df = pd.merge(e_nominal_pcsusie_subset_df, pc_susie_df[['cs_id', 'var_cluster', 'pip', 'pc_num']], on='var_cluster')\n",
    "\n",
    "# pip weighted variance\n",
    "e_nominal_pcsusie_subset_df['variance'] = e_nominal_pcsusie_subset_df['slope'].apply(np.square) * 100\n",
    "e_nominal_pcsusie_subset_df['variance_weighted'] = e_nominal_pcsusie_subset_df['variance'] * e_nominal_pcsusie_subset_df['pip'] \n",
    "\n",
    "# group by phenotype (egene and cluster)\n",
    "sum_variance_df = e_nominal_pcsusie_subset_df.groupby('phenotype_id').agg({'variance_weighted':sum, \n",
    "                                                        'cluster_id':'first', \n",
    "                                                        'variant_id':'first',\n",
    "                                                        'egene_id':'first', \n",
    "                                                        'cs_id':'first'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>end_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>egene_id</th>\n",
       "      <th>var_cluster</th>\n",
       "      <th>var_egene_cluster</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>pc_num</th>\n",
       "      <th>variance</th>\n",
       "      <th>variance_weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13_e_ENSG00...</td>\n",
       "      <td>chr1_966179_G_A_b38</td>\n",
       "      <td>5592</td>\n",
       "      <td>-318</td>\n",
       "      <td>0.838509</td>\n",
       "      <td>129</td>\n",
       "      <td>156</td>\n",
       "      <td>3.469805e-24</td>\n",
       "      <td>0.053869</td>\n",
       "      <td>0.004979</td>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13</td>\n",
       "      <td>ENSG00000187961.13</td>\n",
       "      <td>chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...</td>\n",
       "      <td>chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...</td>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13_1</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>1</td>\n",
       "      <td>0.290188</td>\n",
       "      <td>0.290181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13_e_ENSG00...</td>\n",
       "      <td>chr1_966179_G_A_b38</td>\n",
       "      <td>5592</td>\n",
       "      <td>-318</td>\n",
       "      <td>0.838509</td>\n",
       "      <td>129</td>\n",
       "      <td>156</td>\n",
       "      <td>2.024204e-23</td>\n",
       "      <td>0.053055</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13</td>\n",
       "      <td>ENSG00000187583.10</td>\n",
       "      <td>chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...</td>\n",
       "      <td>chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...</td>\n",
       "      <td>ENSG00000187583.10_ENSG00000187961.13_1</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>1</td>\n",
       "      <td>0.281480</td>\n",
       "      <td>0.281473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...</td>\n",
       "      <td>chr1_1497758_C_T_b38</td>\n",
       "      <td>25989</td>\n",
       "      <td>-14393</td>\n",
       "      <td>0.010352</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.051561e-47</td>\n",
       "      <td>-0.218986</td>\n",
       "      <td>0.013205</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13</td>\n",
       "      <td>ENSG00000160072.19</td>\n",
       "      <td>chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_1</td>\n",
       "      <td>0.247937</td>\n",
       "      <td>1</td>\n",
       "      <td>4.795481</td>\n",
       "      <td>1.188976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...</td>\n",
       "      <td>chr1_1497758_C_T_b38</td>\n",
       "      <td>25989</td>\n",
       "      <td>-14393</td>\n",
       "      <td>0.010352</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>4.022739e-01</td>\n",
       "      <td>-0.014278</td>\n",
       "      <td>0.017029</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13</td>\n",
       "      <td>ENSG00000197785.13</td>\n",
       "      <td>chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_1</td>\n",
       "      <td>0.247937</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020386</td>\n",
       "      <td>0.005054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...</td>\n",
       "      <td>chr1_1499000_C_A_b38</td>\n",
       "      <td>27231</td>\n",
       "      <td>-13151</td>\n",
       "      <td>0.010352</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.051561e-47</td>\n",
       "      <td>-0.218986</td>\n",
       "      <td>0.013205</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13</td>\n",
       "      <td>ENSG00000160072.19</td>\n",
       "      <td>chr1_1499000_C_A_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>chr1_1499000_C_A_b38_ENSG00000160072.19_ENSG00...</td>\n",
       "      <td>ENSG00000160072.19_ENSG00000197785.13_1</td>\n",
       "      <td>0.247937</td>\n",
       "      <td>1</td>\n",
       "      <td>4.795481</td>\n",
       "      <td>1.188976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32327</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...</td>\n",
       "      <td>chr22_50535371_C_G_b38</td>\n",
       "      <td>5315</td>\n",
       "      <td>2791</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>4.031875e-05</td>\n",
       "      <td>-0.053479</td>\n",
       "      <td>0.012885</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.101840</td>\n",
       "      <td>1</td>\n",
       "      <td>0.286000</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32328</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>9.602827e-04</td>\n",
       "      <td>-0.040343</td>\n",
       "      <td>0.012130</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.088869</td>\n",
       "      <td>1</td>\n",
       "      <td>0.162755</td>\n",
       "      <td>0.014464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32329</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>1.981092e-05</td>\n",
       "      <td>-0.051906</td>\n",
       "      <td>0.012023</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.088869</td>\n",
       "      <td>1</td>\n",
       "      <td>0.269419</td>\n",
       "      <td>0.023943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32330</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>6.250424e-04</td>\n",
       "      <td>-0.044693</td>\n",
       "      <td>0.012966</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.101840</td>\n",
       "      <td>1</td>\n",
       "      <td>0.199749</td>\n",
       "      <td>0.020342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32331</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>4.031875e-05</td>\n",
       "      <td>-0.053479</td>\n",
       "      <td>0.012885</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.101840</td>\n",
       "      <td>1</td>\n",
       "      <td>0.286000</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32332 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            phenotype_id  \\\n",
       "0      ENSG00000187583.10_ENSG00000187961.13_e_ENSG00...   \n",
       "1      ENSG00000187583.10_ENSG00000187961.13_e_ENSG00...   \n",
       "2      ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...   \n",
       "3      ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...   \n",
       "4      ENSG00000160072.19_ENSG00000197785.13_e_ENSG00...   \n",
       "...                                                  ...   \n",
       "32327  ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...   \n",
       "32328  ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...   \n",
       "32329  ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...   \n",
       "32330  ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...   \n",
       "32331  ENSG00000025708.13_ENSG00000177989.13_e_ENSG00...   \n",
       "\n",
       "                   variant_id  start_distance  end_distance        af  \\\n",
       "0         chr1_966179_G_A_b38            5592          -318  0.838509   \n",
       "1         chr1_966179_G_A_b38            5592          -318  0.838509   \n",
       "2        chr1_1497758_C_T_b38           25989        -14393  0.010352   \n",
       "3        chr1_1497758_C_T_b38           25989        -14393  0.010352   \n",
       "4        chr1_1499000_C_A_b38           27231        -13151  0.010352   \n",
       "...                       ...             ...           ...       ...   \n",
       "32327  chr22_50535371_C_G_b38            5315          2791  0.021739   \n",
       "32328  chr22_50540263_G_A_b38           10207          7683  0.022774   \n",
       "32329  chr22_50540263_G_A_b38           10207          7683  0.022774   \n",
       "32330  chr22_50541096_C_A_b38           11040          8516  0.021739   \n",
       "32331  chr22_50541096_C_A_b38           11040          8516  0.021739   \n",
       "\n",
       "       ma_samples  ma_count  pval_nominal     slope  slope_se  \\\n",
       "0             129       156  3.469805e-24  0.053869  0.004979   \n",
       "1             129       156  2.024204e-23  0.053055  0.005000   \n",
       "2              10        10  1.051561e-47 -0.218986  0.013205   \n",
       "3              10        10  4.022739e-01 -0.014278  0.017029   \n",
       "4              10        10  1.051561e-47 -0.218986  0.013205   \n",
       "...           ...       ...           ...       ...       ...   \n",
       "32327          20        21  4.031875e-05 -0.053479  0.012885   \n",
       "32328          20        22  9.602827e-04 -0.040343  0.012130   \n",
       "32329          20        22  1.981092e-05 -0.051906  0.012023   \n",
       "32330          20        21  6.250424e-04 -0.044693  0.012966   \n",
       "32331          20        21  4.031875e-05 -0.053479  0.012885   \n",
       "\n",
       "                                  cluster_id            egene_id  \\\n",
       "0      ENSG00000187583.10_ENSG00000187961.13  ENSG00000187961.13   \n",
       "1      ENSG00000187583.10_ENSG00000187961.13  ENSG00000187583.10   \n",
       "2      ENSG00000160072.19_ENSG00000197785.13  ENSG00000160072.19   \n",
       "3      ENSG00000160072.19_ENSG00000197785.13  ENSG00000197785.13   \n",
       "4      ENSG00000160072.19_ENSG00000197785.13  ENSG00000160072.19   \n",
       "...                                      ...                 ...   \n",
       "32327  ENSG00000025708.13_ENSG00000177989.13  ENSG00000177989.13   \n",
       "32328  ENSG00000025708.13_ENSG00000177989.13  ENSG00000025708.13   \n",
       "32329  ENSG00000025708.13_ENSG00000177989.13  ENSG00000177989.13   \n",
       "32330  ENSG00000025708.13_ENSG00000177989.13  ENSG00000025708.13   \n",
       "32331  ENSG00000025708.13_ENSG00000177989.13  ENSG00000177989.13   \n",
       "\n",
       "                                             var_cluster  \\\n",
       "0      chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...   \n",
       "1      chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...   \n",
       "2      chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...   \n",
       "3      chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...   \n",
       "4      chr1_1499000_C_A_b38_ENSG00000160072.19_ENSG00...   \n",
       "...                                                  ...   \n",
       "32327  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "32328  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32329  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32330  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32331  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                       var_egene_cluster  \\\n",
       "0      chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...   \n",
       "1      chr1_966179_G_A_b38_ENSG00000187583.10_ENSG000...   \n",
       "2      chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...   \n",
       "3      chr1_1497758_C_T_b38_ENSG00000160072.19_ENSG00...   \n",
       "4      chr1_1499000_C_A_b38_ENSG00000160072.19_ENSG00...   \n",
       "...                                                  ...   \n",
       "32327  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "32328  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32329  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32330  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "32331  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                         cs_id       pip  pc_num  variance  \\\n",
       "0      ENSG00000187583.10_ENSG00000187961.13_1  0.999976       1  0.290188   \n",
       "1      ENSG00000187583.10_ENSG00000187961.13_1  0.999976       1  0.281480   \n",
       "2      ENSG00000160072.19_ENSG00000197785.13_1  0.247937       1  4.795481   \n",
       "3      ENSG00000160072.19_ENSG00000197785.13_1  0.247937       1  0.020386   \n",
       "4      ENSG00000160072.19_ENSG00000197785.13_1  0.247937       1  4.795481   \n",
       "...                                        ...       ...     ...       ...   \n",
       "32327  ENSG00000025708.13_ENSG00000177989.13_2  0.101840       1  0.286000   \n",
       "32328  ENSG00000025708.13_ENSG00000177989.13_2  0.088869       1  0.162755   \n",
       "32329  ENSG00000025708.13_ENSG00000177989.13_2  0.088869       1  0.269419   \n",
       "32330  ENSG00000025708.13_ENSG00000177989.13_2  0.101840       1  0.199749   \n",
       "32331  ENSG00000025708.13_ENSG00000177989.13_2  0.101840       1  0.286000   \n",
       "\n",
       "       variance_weighted  \n",
       "0               0.290181  \n",
       "1               0.281473  \n",
       "2               1.188976  \n",
       "3               0.005054  \n",
       "4               1.188976  \n",
       "...                  ...  \n",
       "32327           0.029126  \n",
       "32328           0.014464  \n",
       "32329           0.023943  \n",
       "32330           0.020342  \n",
       "32331           0.029126  \n",
       "\n",
       "[32332 rows x 19 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e_nominal_pcsusie_subset_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variance_weighted</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>variant_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenotype_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000001460.17_ENSG00000001461.16_e_ENSG00000001460.17</th>\n",
       "      <td>0.163261</td>\n",
       "      <td>ENSG00000001460.17_ENSG00000001461.16</td>\n",
       "      <td>chr1_24385859_T_A_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001460.17_ENSG00000001461.16_e_ENSG00000001461.16</th>\n",
       "      <td>0.243458</td>\n",
       "      <td>ENSG00000001460.17_ENSG00000001461.16</td>\n",
       "      <td>chr1_24385859_T_A_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000001561.6</th>\n",
       "      <td>0.372927</td>\n",
       "      <td>ENSG00000001561.6_ENSG00000112796.9</td>\n",
       "      <td>chr6_46102670_C_T_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000112796.9</th>\n",
       "      <td>0.199345</td>\n",
       "      <td>ENSG00000001561.6_ENSG00000112796.9</td>\n",
       "      <td>chr6_46102670_C_T_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002726.20_ENSG00000002933.7_ENSG00000055118.14_ENSG00000106565.17_e_ENSG00000002726.20</th>\n",
       "      <td>0.039052</td>\n",
       "      <td>ENSG00000002726.20_ENSG00000002933.7_ENSG00000...</td>\n",
       "      <td>chr7_150780504_T_G_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000274750.2</th>\n",
       "      <td>0.384807</td>\n",
       "      <td>ENSG00000274750.2_ENSG00000275713.2</td>\n",
       "      <td>chr6_26208413_C_G_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000275713.2</th>\n",
       "      <td>0.022816</td>\n",
       "      <td>ENSG00000274750.2_ENSG00000275713.2</td>\n",
       "      <td>chr6_26208413_C_G_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275464.4_ENSG00000280071.3_ENSG00000280433.1_e_ENSG00000275464.4</th>\n",
       "      <td>0.518858</td>\n",
       "      <td>ENSG00000275464.4_ENSG00000280071.3_ENSG000002...</td>\n",
       "      <td>chr21_5089937_T_C_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275464.4_ENSG00000280071.3_ENSG00000280433.1_e_ENSG00000280071.3</th>\n",
       "      <td>0.727659</td>\n",
       "      <td>ENSG00000275464.4_ENSG00000280071.3_ENSG000002...</td>\n",
       "      <td>chr21_5089937_T_C_b38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275464.4_ENSG00000280071.3_ENSG00000280433.1_e_ENSG00000280433.1</th>\n",
       "      <td>0.538699</td>\n",
       "      <td>ENSG00000275464.4_ENSG00000280071.3_ENSG000002...</td>\n",
       "      <td>chr21_5089937_T_C_b38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1335 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    variance_weighted  \\\n",
       "phenotype_id                                                            \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...           0.163261   \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...           0.243458   \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...           0.372927   \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...           0.199345   \n",
       "ENSG00000002726.20_ENSG00000002933.7_ENSG000000...           0.039052   \n",
       "...                                                               ...   \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...           0.384807   \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...           0.022816   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...           0.518858   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...           0.727659   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...           0.538699   \n",
       "\n",
       "                                                                                           cluster_id  \\\n",
       "phenotype_id                                                                                            \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...              ENSG00000001460.17_ENSG00000001461.16   \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...              ENSG00000001460.17_ENSG00000001461.16   \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...                ENSG00000001561.6_ENSG00000112796.9   \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...                ENSG00000001561.6_ENSG00000112796.9   \n",
       "ENSG00000002726.20_ENSG00000002933.7_ENSG000000...  ENSG00000002726.20_ENSG00000002933.7_ENSG00000...   \n",
       "...                                                                                               ...   \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...                ENSG00000274750.2_ENSG00000275713.2   \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...                ENSG00000274750.2_ENSG00000275713.2   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...  ENSG00000275464.4_ENSG00000280071.3_ENSG000002...   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...  ENSG00000275464.4_ENSG00000280071.3_ENSG000002...   \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...  ENSG00000275464.4_ENSG00000280071.3_ENSG000002...   \n",
       "\n",
       "                                                                variant_id  \n",
       "phenotype_id                                                                \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...   chr1_24385859_T_A_b38  \n",
       "ENSG00000001460.17_ENSG00000001461.16_e_ENSG000...   chr1_24385859_T_A_b38  \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...   chr6_46102670_C_T_b38  \n",
       "ENSG00000001561.6_ENSG00000112796.9_e_ENSG00000...   chr6_46102670_C_T_b38  \n",
       "ENSG00000002726.20_ENSG00000002933.7_ENSG000000...  chr7_150780504_T_G_b38  \n",
       "...                                                                    ...  \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...   chr6_26208413_C_G_b38  \n",
       "ENSG00000274750.2_ENSG00000275713.2_e_ENSG00000...   chr6_26208413_C_G_b38  \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...   chr21_5089937_T_C_b38  \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...   chr21_5089937_T_C_b38  \n",
       "ENSG00000275464.4_ENSG00000280071.3_ENSG0000028...   chr21_5089937_T_C_b38  \n",
       "\n",
       "[1335 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_variance_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I now have a table with a row for each pcqtl for the the % var explained for each egene in its cluster. I want to further annotate this with relevant info, like the qtl lead variant to egene tss distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in the gene information (start and strand are what I need)\n",
    "full_gencode=pd.read_csv('/home/klawren/oak/pcqtls/data/references/processed_gencode.v26.GRCh38.genes.gtf', sep='\\t', skiprows=range(6), \n",
    "            header=None, names=['chr', 'dataset', 'type', 'start','end', '.', 'strand', 'na', 'info'])\n",
    "\n",
    "full_gencode = full_gencode[full_gencode['type']=='transcript']\n",
    "full_gencode['transcript_id'] = full_gencode['info'].str.split(';').str[1].str.split('\\\"').str[-2]\n",
    "\n",
    "# add in the start and end info\n",
    "full_gencode['tss_start'] = np.where(full_gencode['strand'] == '+', full_gencode['start'], full_gencode['end'])\n",
    "full_gencode['gene_end'] = np.where(full_gencode['strand'] == '-', full_gencode['start'], full_gencode['end'])\n",
    "\n",
    "# filter to just the transcripts that are in the clusters\n",
    "gene_ids = np.concatenate(overlap_df['cluster_id'].str.split('_'))\n",
    "gid_gencode = full_gencode.set_index('transcript_id').loc[gene_ids]\n",
    "gid_gencode = gid_gencode.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = sum_variance_df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'egene_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3653\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3652\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'egene_id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m egene_tss \u001b[38;5;241m=\u001b[39m gid_gencode\u001b[38;5;241m.\u001b[39mloc[\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43megene_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtss_start\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/series.py:1007\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1004\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m   1006\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1007\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_hashable(key):\n\u001b[1;32m   1010\u001b[0m     \u001b[38;5;66;03m# Otherwise index.get_value will raise InvalidIndexError\u001b[39;00m\n\u001b[1;32m   1011\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1012\u001b[0m         \u001b[38;5;66;03m# For labels that don't resolve as scalars like tuples and frozensets\u001b[39;00m\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/series.py:1116\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1113\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1115\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1116\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1118\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1119\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3655\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3655\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3656\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3657\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3658\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3659\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3660\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'egene_id'"
     ]
    }
   ],
   "source": [
    "egene_tss = gid_gencode.loc[row['egene_id']]['tss_start']\n",
    "lead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
