{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manhatten Plot for example cluster\n",
    "\n",
    "including:\n",
    "* for each egene and cluster pc:\n",
    "    * nominal ps\n",
    "    * lead variant for each independent signal via susie \n",
    "    * credible set via susie\n",
    "* gene annotations\n",
    "    * genes in cluster\n",
    "        * variance in gene expression explained by each pc\n",
    "    * genes not in cluster?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import matplotlib.patches as patches\n",
    "from scipy.stats import linregress\n",
    "\n",
    "\n",
    "import seaborn as sns \n",
    "from tqdm.auto import tqdm  # for notebooks\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.lines as mlines\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib.cm import register_cmap\n",
    "\n",
    "\n",
    "import matplotlib as mpl\n",
    "#mpl.rcParams['figure.dpi']= 700\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# get outputs from a config file\n",
    "prefix = '/home/klawren/oak/pcqtls'\n",
    "import yaml\n",
    "config_path= f'{prefix}/config/pcqtl_proteincoding.yaml'\n",
    "with open(config_path, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "tissue_id_path = config['tissue_id_path']\n",
    "eqtl_output_dir = config['eqtl_output_dir']\n",
    "pcqtl_output_dir = config['pcqtl_output_dir']\n",
    "pc_output_dir = config['pc_output_dir']\n",
    "filtered_expression_output_dir = config['filtered_expression_output_dir']\n",
    "overlap_output_dir = config['overlap_output_dir']\n",
    "\n",
    "\n",
    "# load in the tissue ids \n",
    "tissue_df = pd.read_csv(f\"{prefix}/{tissue_id_path}\", header=0)\n",
    "tissue_ids = list(tissue_df['Tissue'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "decide which cluster to do"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_dfs=[]\n",
    "for i in range(len(tissue_ids)):\n",
    "    tissue_id = tissue_ids[i]\n",
    "    overlap_df = pd.read_csv(f'{prefix}/{overlap_output_dir}/{tissue_id}.v8.overlap.txt', sep='\\t')\n",
    "    overlap_dfs.append(overlap_df)\n",
    "\n",
    "full_overlap_df = pd.concat(overlap_dfs, keys=tissue_ids, names=['Tissue']).reset_index()\n",
    "full_overlap_df['n_genes'] = full_overlap_df['cluster_id'].str.split('_').apply(len)\n",
    "full_overlap_df['pc_num'] = np.where(full_overlap_df['orig_cs_dataset']=='pc_qtl',  full_overlap_df['cs_full_id'].str.split('_').str[-2].str.strip('pc'), 0)\n",
    "full_overlap_df['pc_num'] = full_overlap_df['pc_num'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_df = pd.read_csv('/home/klawren/oak/pcqtls/data/references/ribeiro_2021_coloc_results.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_signals = full_overlap_df[full_overlap_df['lead_variant_id'].str[:-4].isin(coloc_df['variant_id'])]\n",
    "shared_signals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_shared_signals = shared_signals[shared_signals['num_e_overlap'] == 0]\n",
    "filtered_shared_signals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = filtered_shared_signals[filtered_shared_signals['n_genes']> 2].iloc[3]\n",
    "filtered_shared_signals[filtered_shared_signals['n_genes'] >2]\n",
    "\n",
    "# ENSG00000120071.13_ENSG00000176681.14 has really long LD\n",
    "# ENSG00000026950.16_ENSG00000111801.1 is ok but has lots of signals, seems like LD error \n",
    "# ENSG00000140474.12_ENSG00000140497.16 is the one I will do\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_id = row.Tissue\n",
    "chr_id = row['lead_variant_id'].split('_')[0].split('chr')[1]\n",
    "cluster_id = row['cluster_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HLA cluster?\n",
    "#cluster_id = 'ENSG00000204252.13_ENSG00000204257.14_ENSG00000223865.10_ENSG00000231389.7'\n",
    "#chr_id = 6\n",
    "#tissue_id = 'Nerve_Tibial'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# skin cluster \n",
    "#tissue_id = 'Skin_Sun_Exposed_Lower_leg'\n",
    "#chr_id = 17\n",
    "#cluster_id = 'ENSG00000006059.3_ENSG00000094796.4_ENSG00000108244.16_ENSG00000108417.3_ENSG00000108759.3_ENSG00000126337.13_ENSG00000131737.5_ENSG00000131738.10_ENSG00000171360.3_ENSG00000171396.11_ENSG00000171431.3_ENSG00000186860.4_ENSG00000187272.6_ENSG00000188581.8_ENSG00000196156.4_ENSG00000196859.7_ENSG00000197079.8_ENSG00000198083.9_ENSG00000198090.3_ENSG00000198271.4_ENSG00000198443.6_ENSG00000204873.4_ENSG00000204880.7_ENSG00000204889.10_ENSG00000212657.1_ENSG00000212721.3_ENSG00000212722.7_ENSG00000212724.3_ENSG00000212725.3_ENSG00000212899.2_ENSG00000212900.2_ENSG00000212901.3_ENSG00000213416.3_ENSG00000213417.3_ENSG00000214518.3_ENSG00000221852.4_ENSG00000221880.3_ENSG00000240871.5_ENSG00000241595.2_ENSG00000244537.2'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load in the data\n",
    "* nominal p values for control eqtl and cluster\n",
    "* susie for control eqtl and susie\n",
    "* gene annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load nominal p value dfs\n",
    "\n",
    "def var_pos(df):\n",
    "    return df['variant_id'].str.split('_').str[1].astype(int)\n",
    "\n",
    "# load in e nominal\n",
    "def load_e_nominal(path):\n",
    "    e_nominal_df = pd.read_parquet(path)\n",
    "    e_nominal_df['variant_pos'] = var_pos(e_nominal_df)\n",
    "    e_nominal_df['cluster_id'] = e_nominal_df['phenotype_id'].str.split('_e_').str[0]\n",
    "    e_nominal_df['egene_id'] = e_nominal_df['phenotype_id'].str.split('_e_').str[1]\n",
    "    return e_nominal_df\n",
    "\n",
    "e_nominal_df = load_e_nominal(f'{eqtl_output_dir}/{tissue_id}/{tissue_id}.v8.cluster_genes.cis_qtl_pairs.chr{chr_id}.parquet')\n",
    "\n",
    "\n",
    "def load_pc_nominal(path):\n",
    "    pc_nominal_df = pd.read_parquet(path)\n",
    "    pc_nominal_df['variant_pos'] = var_pos(pc_nominal_df)\n",
    "    pc_nominal_df['cluster_id'] = pc_nominal_df['phenotype_id'].str[:-4]\n",
    "    pc_nominal_df['cluster_size'] = pc_nominal_df['phenotype_id'].str.split('_').str.len() - 1\n",
    "    pc_nominal_df['pc_id'] = pc_nominal_df['phenotype_id'].str.split('_pc').str[-1].astype(int)\n",
    "    return pc_nominal_df\n",
    "\n",
    "pc_nominal_df = load_pc_nominal(f'{pcqtl_output_dir}/{tissue_id}/{tissue_id}.v8.pcs.cis_qtl_pairs.chr{chr_id}.parquet')\n",
    "pc_nominal_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load susie dfs\n",
    "def get_lead_var(susie_df):\n",
    "    lv = susie_df.loc[susie_df.groupby('cs_full_id')['pip'].idxmax(),['cs_full_id','variant_id']].set_index('cs_full_id')\n",
    "    return susie_df['cs_full_id'].map(lv['variant_id'])\n",
    "\n",
    "def load_e_susie(path):\n",
    "    e_susie_df = pd.read_csv(path, sep='\\t', index_col=0)\n",
    "    e_susie_df['cluster_id'] = e_susie_df['phenotype_id'].str.split('_e_').str[0]\n",
    "    e_susie_df['egene_id'] = e_susie_df['phenotype_id'].str.split('_e_').str[1]\n",
    "    e_susie_df['cs_full_id'] = e_susie_df['phenotype_id'].astype(str) + '_e_cs' + e_susie_df['cs_id'].astype(str) \n",
    "    e_susie_df['lead_variant'] = get_lead_var(e_susie_df)\n",
    "    e_susie_df['is_lead'] = e_susie_df['variant_id'] == e_susie_df['lead_variant']\n",
    "    e_susie_df['variant_pos'] = var_pos(e_susie_df)\n",
    "    return e_susie_df\n",
    "\n",
    "e_susie_df = load_e_susie(f'{eqtl_output_dir}/{tissue_id}/{tissue_id}.v8.cluster_genes.susie.txt')\n",
    "# filter to only the chromosome we're considering\n",
    "e_susie_df = e_susie_df[e_susie_df['variant_id'].str.split('_').str[0] == f'chr{chr_id}']\n",
    "\n",
    "def load_pc_susie(path):\n",
    "    pc_susie_df = pd.read_csv(path, sep='\\t', index_col=0)\n",
    "    pc_susie_df['cluster_id'] = pc_susie_df['phenotype_id'].str.split('_pc').str[0]\n",
    "    pc_susie_df['cs_full_id'] = pc_susie_df['phenotype_id'].astype(str) + '_cs' + pc_susie_df['cs_id'].astype(str)\n",
    "    pc_susie_df['lead_variant'] = get_lead_var(pc_susie_df)\n",
    "    pc_susie_df['is_lead'] = pc_susie_df['variant_id'] == pc_susie_df['lead_variant']\n",
    "    pc_susie_df['variant_pos'] = var_pos(pc_susie_df)\n",
    "    pc_susie_df['pc_id'] = pc_susie_df['phenotype_id'].str.split('_pc').str[-1].astype(int)\n",
    "    return pc_susie_df\n",
    "\n",
    "pc_susie_df = load_pc_susie(f'{pcqtl_output_dir}/{tissue_id}/{tissue_id}.v8.pcs.susie.txt')\n",
    "# filter to only the chromosome we're considering\n",
    "pc_susie_df = pc_susie_df[pc_susie_df['variant_id'].str.split('_').str[0] == f'chr{chr_id}']\n",
    "\n",
    "pc_susie_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in pc data\n",
    "pc_df = pd.read_csv(f'{prefix}/{pc_output_dir}/{tissue_id}.pcs.bed', sep='\\t')\n",
    "# get a list of all the sample ids\n",
    "sample_ids = pc_df.columns[4:]\n",
    "# add a cluster id and pc number column\n",
    "pc_df['cluster_id'] = pc_df['gene_id'].str.split('_pc').str[0]\n",
    "pc_df['pc_id'] = pc_df['gene_id'].str.split('_pc').str[1].astype('float')\n",
    "pc_df['cluster_size'] = pc_df['cluster_id'].str.split('_').apply(len)\n",
    "# filter to only the chromosome we're considering\n",
    "pc_df = pc_df[pc_df['#chr'] == f'chr{chr_id}']\n",
    "\n",
    "pc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in gene expression data\n",
    "expression_df = pd.read_csv(f'{prefix}/{filtered_expression_output_dir}/{tissue_id}.v8.normalized_expression.cluster_genes.bed', sep='\\t')\n",
    "# add a cluster id and pc number column\n",
    "expression_df['cluster_id'] = expression_df['gene_id'].str.split('_e_').str[0]\n",
    "expression_df['egene_id'] = expression_df['gene_id'].str.split('_e_').str[1]\n",
    "# filter to only the chromosome we're considering\n",
    "expression_df = expression_df[expression_df['#chr'] == f'chr{chr_id}']\n",
    "\n",
    "\n",
    "expression_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load gene annotations\n",
    "full_gencode=pd.read_csv('/home/klawren/oak/pcqtls/data/references/processed_gencode.v26.GRCh38.genes.gtf', sep='\\t', skiprows=range(6), \n",
    "            header=None, names=['chr', 'dataset', 'type', 'start','end', '.', 'strand', 'na', 'info'])\n",
    "\n",
    "gencode = full_gencode[full_gencode['type']=='transcript']\n",
    "gencode['transcript_id'] = gencode['info'].str.split(';').str[1].str.split('\\\"').str[-2]\n",
    "gencode['gene_name'] = gencode['info'].str.split(';').str[3].str.split('\\\"').str[-2]\n",
    "gencode['tss_start'] = np.where(gencode['strand'] == '+', gencode['start'], gencode['end'])\n",
    "gencode['gene_end'] = np.where(gencode['strand'] == '-', gencode['start'], gencode['end'])\n",
    "\n",
    "\n",
    "# filter to just the genes in clusters on the chromosome we're considering\n",
    "gene_ids = expression_df['egene_id'].unique()\n",
    "gid_gencode = gencode.set_index('transcript_id').loc[gene_ids]\n",
    "gid_gencode = gid_gencode.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the gene-gene correlations\n",
    "\n",
    "def plot_cluster_corr(cluster_id, expression_df, sample_ids, gid_gencode=gid_gencode):\n",
    "    # get the gene ids (this will be the order on the heatmap)\n",
    "    gene_ids = cluster_id.split('_')\n",
    "    # set the order to be the order along the genome\n",
    "    cluster_gencode = gid_gencode.loc[gene_ids]\n",
    "    sorted_gene_ids = cluster_gencode.sort_values(by=['start', 'end']).index.values\n",
    "\n",
    "\n",
    "    # pull out expression for this cluster\n",
    "    expression_cluster = expression_df[expression_df['cluster_id'] == cluster_id]\n",
    "    # get the expression in the same order as the gene ids\n",
    "    expression_cluster = expression_cluster.set_index('egene_id').loc[sorted_gene_ids]\n",
    "    # get correlations \n",
    "    corrs = expression_cluster[sample_ids].T.corr(method='spearman')\n",
    "\n",
    "    # rename with gene names\n",
    "    gene_names = cluster_gencode.sort_values(by=['start', 'end'])['gene_name']\n",
    "    corrs = corrs.rename(index=gene_names, columns=gene_names)\n",
    "\n",
    "    # make a plot\n",
    "    fig,ax = plt.subplots(figsize=(7,6))\n",
    "    # mask so the off diagonal isn't shown\n",
    "    mask = np.triu(np.ones_like(corrs)) \n",
    "    # create the colormap, these are colors matched to the manhattenplots. coolwarm is another good option\n",
    "    cmap = LinearSegmentedColormap.from_list('mycmap', [(0, '#c4553aff'), (.5, 'white'), (1, '#3e8093ff')])\n",
    "    # plot the correlations\n",
    "    sns.heatmap(corrs, mask=mask, cmap=cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=True, yticklabels=True)\n",
    "\n",
    "    # rotate the gene names\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "    ax.set_yticklabels(ax.get_yticklabels(), rotation=0)\n",
    "\n",
    "    # no axis labels\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_xlabel('')\n",
    "\n",
    "\n",
    "    # removes the top and right x-tick labels \n",
    "    plt.xticks()[1][-1].set_visible(False)\n",
    "    plt.yticks()[1][0].set_visible(False)\n",
    "\n",
    "    fig.tight_layout()\n",
    "\n",
    "\n",
    "plot_cluster_corr(cluster_id, expression_df, sample_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the gene-gene correlations for one of the big keratin clusters\n",
    "\n",
    "# load in gene expression data\n",
    "skin_expression_df = pd.read_csv(f'{prefix}/{filtered_expression_output_dir}/Skin_Not_Sun_Exposed_Suprapubic.v8.normalized_expression.cluster_genes.bed', sep='\\t')\n",
    "skin_sample_ids = skin_expression_df.columns[4:]\n",
    "# add a cluster id and pc number column\n",
    "skin_expression_df['cluster_id'] = skin_expression_df['gene_id'].str.split('_e_').str[0]\n",
    "skin_expression_df['egene_id'] = skin_expression_df['gene_id'].str.split('_e_').str[1]\n",
    "\n",
    "skin_cluster_id = 'ENSG00000006059.3_ENSG00000108244.16_ENSG00000131737.5_ENSG00000131738.10_ENSG00000167916.4_ENSG00000167920.8_ENSG00000171396.11_ENSG00000171431.3_ENSG00000171446.6_ENSG00000173908.8_ENSG00000180386.7_ENSG00000186393.5_ENSG00000186395.7_ENSG00000186860.4_ENSG00000187242.4_ENSG00000187272.6_ENSG00000188581.8_ENSG00000196156.4_ENSG00000196859.7_ENSG00000198083.9_ENSG00000198090.3_ENSG00000198271.4_ENSG00000198443.6_ENSG00000204873.4_ENSG00000204880.7_ENSG00000204889.10_ENSG00000204897.6_ENSG00000212657.1_ENSG00000212658.1_ENSG00000212659.1_ENSG00000212721.3_ENSG00000212722.7_ENSG00000212724.3_ENSG00000212725.3_ENSG00000212899.2_ENSG00000212900.2_ENSG00000212901.3_ENSG00000213416.3_ENSG00000213417.3_ENSG00000214518.3_ENSG00000221852.4_ENSG00000221880.3_ENSG00000239886.5_ENSG00000240871.5_ENSG00000241595.2_ENSG00000244537.2'\n",
    "\n",
    "plot_cluster_corr(skin_cluster_id,skin_expression_df, skin_sample_ids, gid_gencode=gencode.set_index('transcript_id').drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row.lead_variant_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the cluster we're considering\n",
    "full_overlap_df['pc_num'] = np.where(full_overlap_df['orig_cs_dataset']=='pc_qtl',  full_overlap_df['cs_full_id'].str.split('_').str[-2].str.strip('pc'), 0)\n",
    "full_overlap_df['pc_num'] = full_overlap_df['pc_num'].astype(int)\n",
    "full_overlap_df[(full_overlap_df['Tissue']==tissue_id)&(full_overlap_df['cluster_id'] == cluster_id)]\n",
    "\n",
    "# chr15_74932469_T_A_b38\n",
    "# chr6_26372599_C_T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cluster(cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_pos = int(row.lead_variant_id.split('_')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get only variants near the top snp (1kb)\n",
    "pc_nominal_set['var_position'] = pc_nominal_set['variant_id'].str.split('_').str[1].astype(int)\n",
    "pc_nominal_set[abs(pc_nominal_set['var_position'] - var_pos) < 10000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pc values for the pc tahthas teh gwas hit var\n",
    "pc_nominal_set = pc_nominal_df[(pc_nominal_df['cluster_id'] == cluster_id)&((pc_nominal_df['pc_id'] == row.pc_num))]\n",
    "# get only variants near the top snp (10kb)\n",
    "pc_nominal_set['var_position'] = pc_nominal_set['variant_id'].str.split('_').str[1].astype(int)\n",
    "pc_nominal_set = pc_nominal_set[abs(pc_nominal_set['var_position'] - var_pos) < 10000]\n",
    "\n",
    "pc_susie_set=pc_susie_df[(pc_susie_df['cluster_id'] == cluster_id)&(pc_susie_df['pc_id']==row.pc_num)]\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "for i in range(len(cluster_id.split('_'))):\n",
    "    # p values and credible set for this egene\n",
    "    nominal_set = e_nominal_df[e_nominal_df['egene_id'] == cluster_id.split('_')[i]]\n",
    "    # get only variants near the top snp (10kb)\n",
    "    nominal_set['var_position'] = nominal_set['variant_id'].str.split('_').str[1].astype(int)\n",
    "    nominal_set = nominal_set[abs(nominal_set['var_position'] - var_pos) < 10000]\n",
    "    \n",
    "    susie_set=e_susie_df[e_susie_df['egene_id'] == cluster_id.split('_')[i]]\n",
    "\n",
    "    # merge into 1 df for plotting\n",
    "    merged_df = pd.merge(pc_nominal_set, nominal_set, on='variant_id', suffixes=('_pc', '_e'))\n",
    "\n",
    "    # nominal p values\n",
    "    sns.scatterplot(merged_df, x='pval_nominal_pc', y='pval_nominal_e',\n",
    "                            s=10, \n",
    "                            linewidth=0, label=gid_gencode.loc[cluster_id.split('_')[i]]['gene_name'], ax=ax)\n",
    "\n",
    "ax.set(yscale='log', xscale='log')\n",
    "ax.invert_yaxis()\n",
    "ax.invert_xaxis()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cluster(cluster_id, x_padding=400000)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot to show how the e gene and pc1 qtl coloc??\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## split out types of plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genes(ax, cluster_id, scalebar=True, x_padding = 1000000, palette = sns.color_palette()):\n",
    "\n",
    "    # get the gene ids, assumes cluster id is sorted by position \n",
    "    gene_ids = cluster_id.split('_')\n",
    "    # set the order to be the order along the genome\n",
    "    cluster_gencode = gid_gencode.loc[gene_ids]\n",
    "\n",
    "    # Sort the dataframe by 'start'\n",
    "    cluster_gencode = cluster_gencode.sort_values(by=['start', 'end'])\n",
    "\n",
    "    # this list will store the end of the last gene at each level:\n",
    "    ends_of_genes = [0]\n",
    "    # row height:\n",
    "    height = .3\n",
    "\n",
    "    # for each gene\n",
    "    for i in range(len(cluster_gencode)):\n",
    "        gene = cluster_gencode.iloc[i]\n",
    "        # for each level\n",
    "        for level, end in enumerate(ends_of_genes):\n",
    "            # if this gene does not overlap with the last gene at this level\n",
    "            if gene['start'] > end:\n",
    "                # place the gene at this level\n",
    "                arrow = patches.FancyArrow(x = gene['tss_start'], y=height*level, dx=(gene['gene_end'] - gene['tss_start']), \n",
    "                                           dy=0, width=.1, head_length = .02*x_padding, edgecolor=palette[i%10], \n",
    "                                           facecolor=palette[i%10])\n",
    "\n",
    "                ax.add_patch(arrow)\n",
    "                # update the end of the last gene at this level\n",
    "                ends_of_genes[level] = gene['end']\n",
    "                break\n",
    "        else:\n",
    "            # all levels are occupied, we need to create a new level:\n",
    "            ends_of_genes.append(gene['gene_end'])\n",
    "            arrow = patches.FancyArrow(x = gene['tss_start'], y=height*(len(ends_of_genes) - 1), dx=(gene['gene_end'] - gene['tss_start']), \n",
    "                                           dy=0, width=.1, head_length = .02*x_padding, edgecolor=palette[i%10], \n",
    "                                           facecolor=palette[i%10])\n",
    "\n",
    "            ax.add_patch(arrow)\n",
    "    \n",
    "    # Set plot limits\n",
    "    ax.set_ylim(-1*height, height*len(ends_of_genes))\n",
    "    # padding for the cis window\n",
    "    ax.set_xlim(cluster_gencode['start'].min() - x_padding, cluster_gencode['end'].max() + x_padding)\n",
    "\n",
    "    # to remove black outlines for gene plot\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    # remove y axis and ticks\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    # remove x axis and ticks\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.set_xticks([])\n",
    "\n",
    "    return ax\n",
    "\n",
    "\n",
    "def plot_scalebar(ax, scalebar_length = 200000, side='left'):\n",
    "    # add in scalebar\n",
    "    scalebar_label=f'{scalebar_length/1000:.0f} kb'\n",
    "\n",
    "    if side=='left':\n",
    "        # Adding scale bar\n",
    "        scalebar_start = ax.get_xlim()[0]+.05*scalebar_length\n",
    "        scalebar = mlines.Line2D([scalebar_start, scalebar_start + scalebar_length], [np.mean(ax.get_ylim()),np.mean(ax.get_ylim())], color='k')\n",
    "    elif side=='right':\n",
    "        scalebar_start = ax.get_xlim()[1]-.05*scalebar_length - scalebar_length\n",
    "        scalebar = mlines.Line2D([scalebar_start, scalebar_start + scalebar_length], [np.mean(ax.get_ylim()),np.mean(ax.get_ylim())], color='k')\n",
    "\n",
    "    ax.add_line(scalebar)\n",
    "\n",
    "    # Adding caps for bar\n",
    "    cap_length = (ax.get_ylim()[1] - ax.get_ylim()[0])/5\n",
    "    cap1 = mlines.Line2D([scalebar_start, scalebar_start], [np.mean(ax.get_ylim()) - cap_length / 2., np.mean(ax.get_ylim()) + cap_length / 2.], color='k')\n",
    "    cap2 = mlines.Line2D([scalebar_start + scalebar_length, scalebar_start + scalebar_length], [np.mean(ax.get_ylim()) - cap_length / 2., np.mean(ax.get_ylim()) + cap_length / 2.], color='k')\n",
    "    ax.add_line(cap1)\n",
    "    ax.add_line(cap2)\n",
    "\n",
    "    # Adding text\n",
    "    ax.text(scalebar_start + scalebar_length / 2, np.mean(ax.get_ylim()) + cap_length, scalebar_label, ha='center')  \n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cs(ax, nominal_set, susie_set, gene_color, use_color_cs=True):\n",
    "        # nominal p values\n",
    "        sns.scatterplot(nominal_set,\n",
    "                        x='variant_pos', \n",
    "                        y='pval_nominal', \n",
    "                        ax=ax,\n",
    "                        color=gene_color, \n",
    "                        s=10, \n",
    "                        linewidth=0)\n",
    "        # susie credible sets\n",
    "        sns.scatterplot(nominal_set[nominal_set['variant_id'].isin(susie_set['variant_id'])],\n",
    "                        x='variant_pos', \n",
    "                        y='pval_nominal', \n",
    "                        ax=ax,\n",
    "                        color=gene_color if use_color_cs else 'black', \n",
    "                        marker='*' if use_color_cs else '.', \n",
    "                        s=100, \n",
    "                        linewidth=.1)\n",
    "        # susie lead vars sets\n",
    "        sns.scatterplot(nominal_set[nominal_set['variant_id'].isin(susie_set[susie_set['is_lead']]['variant_id'])],\n",
    "                        x='variant_pos', \n",
    "                        y='pval_nominal', \n",
    "                        ax=ax,\n",
    "                        color=gene_color if use_color_cs else '#B83A4B', \n",
    "                        marker='x', \n",
    "                        s=100, \n",
    "                        linewidth=3)\n",
    "        \n",
    "        # to remove black outlines for gene plot\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "\n",
    "        # set y label\n",
    "        ax.set_ylabel('nominal p-value')\n",
    "\n",
    "        \n",
    "        return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "        \n",
    "def add_legend(ax, sorted_gene_ids, palette=sns.color_palette(), use_color_cs=True, var_plot=True):\n",
    "    # create handles for 'gene' section of legend\n",
    "    gene_names = gid_gencode.loc[sorted_gene_ids]['gene_name']\n",
    "    handles_colors_genes = [patches.Patch(color=palette[i], label=gene) for i, gene in enumerate(gene_names)]\n",
    "    # create handels for 'pcs' section of legend\n",
    "    handles_colors_pcs = [patches.Patch(color=palette[i+len(sorted_gene_ids)+1], label=f'pc {i+1}') for i in range(len(sorted_gene_ids))]\n",
    "    # create custom handles for 'marks' section of legend\n",
    "    handles_marks = [Line2D([0], [0], color='black', marker='.', markersize=4, linestyle='None', label='nominal p-value'),\n",
    "                    Line2D([0], [0], color='black', marker='*' if use_color_cs else '.', markersize=7, linestyle='None', label='credible set'),\n",
    "                    Line2D([0], [0], color='#B83A4B', marker='x', markersize=7, linestyle='None', label='lead variant', markeredgewidth=3)]\n",
    "    # if var plot \n",
    "    handles_marks.append(Line2D([0], [0], color=palette[0], marker='P', markersize=7, linestyle='None', label='% variance explained', markeredgewidth=0))\n",
    "\n",
    "    # add in section titles\n",
    "    handles = [Line2D([], [], color='none', label='Genes'), *handles_colors_genes, \n",
    "               Line2D([], [], color='none', label='PCs'), *handles_colors_pcs,\n",
    "               Line2D([], [], color='none', label='Markers'), *handles_marks]\n",
    "\n",
    "    ax.legend(handles=handles, loc='center left', frameon=False)\n",
    "\n",
    "    # to remove black outlines for gene plot\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    # remove y axis and ticks\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    # remove x axis and ticks\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.set_xticks([])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rs(cluster_id):\n",
    "    # the expression and pc values for this cluster\n",
    "    expression_cluster = expression_df[expression_df['cluster_id'] == cluster_id].reset_index()\n",
    "    pc_cluster = pc_df[pc_df['cluster_id'] == cluster_id].reset_index()\n",
    "    # pull gene locations \n",
    "    cluster_gencode = gid_gencode.loc[expression_cluster['egene_id']]\n",
    "    rs = np.zeros((len(expression_cluster), len(pc_cluster)))\n",
    "\n",
    "    for i in range(len(expression_cluster)):\n",
    "        for j in range(len(pc_cluster)):\n",
    "            # pull the right data\n",
    "            pc_values = pc_cluster.loc[j, sample_ids].astype('float')\n",
    "            expression_values = expression_cluster.loc[i, sample_ids].astype('float')\n",
    "            # get the r squared value\n",
    "            slope, intercept, r_value, p_value, std_err = linregress(pc_values, expression_values)\n",
    "            rs[i,j] = (r_value**2)*100\n",
    "\n",
    "    rs_df = pd.DataFrame(rs, columns=[f'pc{i+1}' for i in range(len(pc_cluster))], index=expression_cluster['egene_id'])\n",
    "\n",
    "    rs_df = rs_df.loc[cluster_gencode.index]\n",
    "    rs_df.reset_index()\n",
    "\n",
    "    # add the location for the variance to be ploted as the gene tss\n",
    "    rs_df['position'] = (cluster_gencode['start'] + cluster_gencode['end'])/2\n",
    "\n",
    "    return rs_df\n",
    "\n",
    "def plot_pc_var(ax, rs_df, pc_nums, gene_ids, gene_palette, line=False):\n",
    "    plot_rs = rs_df.loc[gene_ids, [*pc_nums, 'position']]\n",
    "\n",
    "    # split out all the pc values into one column\n",
    "    rs_melt_df = plot_rs.reset_index().melt(id_vars=['position', 'transcript_id'], value_vars=plot_rs.columns[:-1].values, \n",
    "                  var_name='pc_num')\n",
    "    \n",
    "    # plot the varaince explained as a scatterplot\n",
    "    sns.lineplot(rs_melt_df, y='value', x='position', style='pc_num', color='k', ax=ax, legend=False, alpha=0.5)\n",
    "    sns.scatterplot(rs_melt_df, y='value', x='position', hue='transcript_id', ax=ax, palette=gene_palette, legend=False, linewidth=0, s=50, marker='P')\n",
    "\n",
    "    ax.set_ylabel('% expression \\nvariance explained')\n",
    "    ax.set_ylim((0,100))\n",
    "    ax.spines['top'].set_visible(False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cluster(cluster_id, x_padding = 1000000, figsize=(20, 15), var_plot=True):\n",
    "    # get gene ids\n",
    "    gene_ids = cluster_id.split('_')\n",
    "    # sort gene ids by start position\n",
    "    sorted_gene_ids = gid_gencode.loc[gene_ids].sort_values(by=['start', 'end']).index.values\n",
    "    gene_names = gid_gencode.loc[sorted_gene_ids]['gene_name']\n",
    "\n",
    "\n",
    "    # create a gridspec object\n",
    "    # Create a 2x2 figure with specified figure width\n",
    "    fig = plt.figure(figsize=figsize) \n",
    "\n",
    "    num_genes = len(gene_ids)\n",
    "    num_manhatten = num_genes*2\n",
    "    # colors (1st half is genes, 2nd half is pcs)\n",
    "    palette = sns.diverging_palette(220, 20, n=num_genes*2+1)\n",
    "\n",
    "    # cols determines the width of the legend vs the plots\n",
    "    num_cols=10\n",
    "\n",
    "    # set up axes for the different plots \n",
    "    gs = gridspec.GridSpec(num_manhatten*2+1, num_cols)\n",
    "\n",
    "    # one plot for each gene's manhatten, plus one for gene arrows on the bottom\n",
    "    # manhatten plots are twice the height as the arrow plot\n",
    "    axes_arrows = plt.subplot(gs[num_manhatten*2, :num_cols-1])\n",
    "    # plot for the legend on the right side\n",
    "    axes_legend = plt.subplot(gs[:, num_cols-1])\n",
    "\n",
    "    # axes for each manhattenplot\n",
    "    for i in range(num_manhatten):\n",
    "        if i==0:\n",
    "            axs = [plt.subplot(gs[i*2:i*2 + 2, :num_cols-1], sharex=axes_arrows)]\n",
    "            # log scale y axis\n",
    "            axs[0].set(yscale='log')\n",
    "            axs[0].invert_yaxis()\n",
    "        else:\n",
    "            axs.append(plt.subplot(gs[i*2:i*2 + 2, :num_cols-1], sharex=axes_arrows, sharey=axs[0]))\n",
    "\n",
    "\n",
    "    # add in the gene arrows\n",
    "    plot_genes(axes_arrows, cluster_id, x_padding=x_padding, palette = palette)\n",
    "    axes_arrows.set_title('Genes')\n",
    "\n",
    "    # add in a scalebar\n",
    "    plot_scalebar(axes_arrows, scalebar_length = x_padding/5, side='right')\n",
    "\n",
    "    # add in manhattenplots for each e gene\n",
    "    for i in range(num_genes):\n",
    "        nominal_set = e_nominal_df[e_nominal_df['egene_id'] == sorted_gene_ids[i]]\n",
    "        susie_set=e_susie_df[e_susie_df['egene_id'] == sorted_gene_ids[i]]\n",
    "        plot_cs( axs[i], nominal_set, susie_set, palette[i], use_color_cs=False)\n",
    "        axs[i].set_title(f'eQTL for {gene_names[i]}')\n",
    "\n",
    "\n",
    "    # add in manhattenplots for each pc\n",
    "    for i in range(num_genes):\n",
    "        nominal_set = pc_nominal_df[(pc_nominal_df['cluster_id'] == cluster_id)&((pc_nominal_df['pc_id'] == i+1))]\n",
    "        susie_set=pc_susie_df[(pc_susie_df['cluster_id'] == cluster_id)&(pc_susie_df['pc_id']==i+1)]\n",
    "        plot_cs(axs[i+num_genes], nominal_set, susie_set, palette[num_genes+i+1], use_color_cs=False)\n",
    "        axs[i+num_genes].set_title(f'pcQTL for pc {i+1}')\n",
    "\n",
    "    if var_plot:\n",
    "        # add in pc variance explained for each gene by pcs (on each pc plot)\n",
    "        rs_df = get_rs(cluster_id)\n",
    "        # add on each pc plot the var explained for only that pc\n",
    "        for i in range(num_genes):\n",
    "            # shared x axis, different y axis\n",
    "            ax_twin = axs[i+num_genes].twinx()\n",
    "            plot_pc_var(ax_twin, rs_df, [f'pc{i+1}'], gene_ids, palette[:num_genes])\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "    # add in a legend\n",
    "    add_legend(axes_legend, sorted_gene_ids, palette=palette, use_color_cs=False, var_plot=var_plot)\n",
    "\n",
    "    # add some spacing\n",
    "    #plt.subplots_adjust(hspace=2)\n",
    "    fig.tight_layout()  # otherwise the right y-label is slightly clipped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pc 1 nominal set\n",
    "pc_nominal_set = pc_nominal_df[(pc_nominal_df['cluster_id'] == cluster_id)&((pc_nominal_df['pc_id'] == 1))]\n",
    "pc_susie_set=pc_susie_df[(pc_susie_df['cluster_id'] == cluster_id)&(pc_susie_df['pc_id']==1)]\n",
    "\n",
    "nominal_set = e_nominal_df[e_nominal_df['egene_id'] == cluster_id.split('_')[0]]\n",
    "susie_set=e_susie_df[e_susie_df['egene_id'] == cluster_id.split('_')[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(len(cluster_id.split('_')), figsize=(4, 12))\n",
    "\n",
    "lead_var = 'chr15_74860799_A_G_b38'\n",
    "cs_var = pc_susie_set[pc_susie_set['lead_variant'] == 'chr15_74860799_A_G_b38']['variant_id']\n",
    "gene_names = gid_gencode.loc[cluster_id.split('_')]['gene_name']\n",
    "palette = sns.diverging_palette(220, 20, n=len(gene_names)*2+1)\n",
    "\n",
    "\n",
    "\n",
    "for i in range(len(cluster_id.split('_'))):\n",
    "    gene_id = cluster_id.split('_')[i]\n",
    "    nominal_set = e_nominal_df[e_nominal_df['egene_id'] == gene_id]\n",
    "    susie_set=e_susie_df[e_susie_df['egene_id'] == gene_id]\n",
    "    merged_df = pd.merge(pc_nominal_set, nominal_set, on='variant_id', suffixes=('_pc', '_e'))\n",
    "\n",
    "    # nominal p values\n",
    "    sns.scatterplot(merged_df, x='pval_nominal_pc', y='pval_nominal_e', ax = axes[i] ,\n",
    "                        color=palette[i], \n",
    "                        s=10, \n",
    "                        linewidth=0)\n",
    "    # susie credible sets\n",
    "    sns.scatterplot(merged_df[merged_df['variant_id'].isin(cs_var)], x='pval_nominal_pc', y='pval_nominal_e', ax = axes[i],\n",
    "                        color='black', \n",
    "                        marker='.', \n",
    "                        s=100, \n",
    "                        linewidth=.1)\n",
    "    # susie lead vars sets\n",
    "    sns.scatterplot(merged_df[merged_df['variant_id'] == lead_var], x='pval_nominal_pc', y='pval_nominal_e', ax = axes[i],\n",
    "                        color= '#B83A4B', \n",
    "                        marker='x', \n",
    "                        s=100, \n",
    "                        linewidth=3)\n",
    "\n",
    "    axes[i].set(yscale='log', xscale='log')\n",
    "    #axes[i].set_ylim(axes[i].get_xlim())\n",
    "    axes[i].invert_yaxis()\n",
    "    axes[i].invert_xaxis()\n",
    "\n",
    "    # to remove black outlines for gene plot\n",
    "    axes[i].spines['top'].set_visible(False)\n",
    "    axes[i].spines['right'].set_visible(False)\n",
    "\n",
    "    axes[i].set_xlabel('pcQTL nominal p-value')\n",
    "    axes[i].set_ylabel(f'{gene_names[i]} eQTL nominal p-value')\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr15_74860799_A_G_b38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    for i in range(num_genes):\n",
    "        nominal_set = e_nominal_df[e_nominal_df['egene_id'] == gene_ids[i]]\n",
    "        susie_set=e_susie_df[e_susie_df['egene_id'] == gene_ids[i]]\n",
    "        plot_cs( axs[i], nominal_set, susie_set, palette[i], use_color_cs=False)\n",
    "        axs[i].set_title(f'eQTL for {gene_names[i]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_df[expression_df['cluster_id'].str.split('_').apply(len) > 3]['cluster_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull an example cluster\n",
    "#cluster_id = expression_df['cluster_id'].loc[369]\n",
    "#cluster_id = 'ENSG00000142089.15_ENSG00000185201.16_ENSG00000185885.15'\n",
    "#cluster_id = expression_df[expression_df['cluster_id'].str.split('_').apply(len) > 3]['cluster_id'].unique()[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cluster(cluster_id, x_padding=100000)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "rs_df = get_rs(cluster_id)\n",
    "\n",
    "\n",
    "plot_pc_var(ax, rs_df, ['pc1', 'pc2'], gene_ids, palette[num_genes+1:], palette[:num_genes])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_nominal_df[(pc_nominal_df['cluster_id'] == cluster_id)&((pc_nominal_df['pc_id'] == i))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    cluster_gencode = gid_gencode.loc[gene_ids]\n",
    "\n",
    "    # Sort the dataframe by 'start'\n",
    "    cluster_gencode = cluster_gencode.sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot \n",
    "add_text=False\n",
    "max_num_pcs = 4\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "# set axis limits \n",
    "ax.set_ylim(-(len(cluster_gencode)+1)*3,100)\n",
    "x_padding = 10000\n",
    "ax.set_xlim(cluster_gencode['start'].min() - x_padding, cluster_gencode['end'].max() + x_padding)\n",
    "\n",
    "\n",
    "# add a rectangle for each gene location\n",
    "for i in range(len(cluster_gencode)):\n",
    "    row = cluster_gencode.iloc[i]\n",
    "    # rectangle patch for gene body\n",
    "    rect = patches.Rectangle((row['start'], -3-i*3), row['end']-row['start'], 2, linewidth=1, edgecolor=sns.color_palette()[i%10], facecolor=sns.color_palette()[i%10])\n",
    "    ax.add_patch(rect)\n",
    "    # black line for tss location\n",
    "    rect = patches.Rectangle((row['tss_start'], -3-i*3), 200, 2, linewidth=1, edgecolor='k', facecolor='k')\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "plot_rs_melt_df = rs_melt_df[rs_melt_df['pc_num'].isin([f'pc{i+1}' for i in range(max_num_pcs)])]\n",
    "\n",
    "# plot the varaince explained as a scatterplot\n",
    "sns.lineplot(plot_rs_melt_df, y='value', x='position', style='pc_num', color='grey')\n",
    "sns.scatterplot(plot_rs_melt_df, y='value', x='position', hue='transcript_id', style='pc_num', ax=ax)\n",
    "\n",
    "if add_text:\n",
    "    # add text to the scatterplot\n",
    "    for i in range(plot_rs_melt_df.shape[0]):\n",
    "        ax.text(plot_rs_melt_df['position'].iloc[i]+100, plot_rs_melt_df['value'].iloc[i]+1, \n",
    "            '{:.2f}%'.format(plot_rs_melt_df['value'].iloc[i]), horizontalalignment='left', size='medium', color='k')\n",
    "\n",
    "# Add the patch to the Axes\n",
    "ax.set_xlabel('Position (bp)')\n",
    "ax.set_ylabel('% variance explained')\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "debuggin get_pcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_dir = config['clusters_dir']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in the clusters data\n",
    "clusters = []\n",
    "for t in tissue_df['Tissue']:\n",
    "    clusters.append(pd.read_csv(f'{prefix}/{clusters_dir}/{t}_clusters_all_chr.csv'))\n",
    "clusters_df = pd.concat(clusters, keys=tissue_df['Tissue'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row=clusters_df.iloc[0]\n",
    "cluster_id ='_'.join([*sorted(row['Transcripts'].split(','))])\n",
    "expression_df = pd.read_csv(f'{prefix}/{filtered_expression_output_dir}/{row.Tissue}.v8.normalized_expression.cluster_genes.bed', sep='\\t')\n",
    "\n",
    "sample_ids = expression_df.columns[4:]\n",
    "\n",
    "expression_df['egene_id'] = expression_df['gene_id'].str.split('_e_').str[1]\n",
    "expression_df['cluster_id'] = expression_df['gene_id'].str.split('_e_').str[0]\n",
    "\n",
    "expression_df_gid = expression_df.set_index('egene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all pcs and combine \n",
    "cluster_expression_df_gid = expression_df_gid[expression_df_gid['cluster_id']==cluster_id]\n",
    "cluster = cluster_expression_df_gid.loc[row['Transcripts'].split(',')]\n",
    "X = cluster[sample_ids].transpose()\n",
    "pca = PCA()\n",
    "pc_values = pca.fit_transform(X)\n",
    "\n",
    "# get an id for each pc\n",
    "gene_ids = []\n",
    "for pc_num in range(pc_values.shape[1]):\n",
    "    gene_ids.append('_'.join([*sorted(row['Transcripts'].split(',')), f'pc{pc_num+1}']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
