{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 1:  Co-expressed genes share biological function and regulatory architecture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import upsetplot as up\n",
    "from tqdm.auto import tqdm \n",
    "from scipy.stats import spearmanr\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "from sklearn.decomposition import PCA\n",
    "import pyranges as pr\n",
    "\n",
    "# set higher figure resolution\n",
    "import matplotlib as mpl\n",
    "#mpl.rcParams['figure.dpi']= 1000\n",
    "\n",
    "tqdm.pandas()\n",
    "\n",
    "# get outputs from a config file\n",
    "prefix = '/home/klawren/oak/pcqtls'\n",
    "import yaml\n",
    "config_path= f'{prefix}/config/proteincoding_main.yaml'\n",
    "with open(config_path, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/klawren/oak/pcqtls/workflow/scripts')\n",
    "from notebook_helper_functions import *\n",
    "from annotate_null_clusters import * \n",
    "from residualize import calculate_residual\n",
    "\n",
    "# colors\n",
    "corr_cmap = LinearSegmentedColormap.from_list('corr', [(0, '#3e8093ff'), (.5, 'white'), (1, '#c4553aff')])\n",
    "gtex_tissue_dict = {\n",
    "    \"Adipose_Subcutaneous\": \"#FFA54F\",\n",
    "    \"Adipose_Visceral_Omentum\": \"#EE9A00\",\n",
    "    \"Artery_Tibial\": \"#FF0000\",\n",
    "    \"Cells_Cultured_fibroblasts\": \"#9AC0CD\",\n",
    "    \"Esophagus_Mucosa\": \"#8B7355\",\n",
    "    \"Esophagus_Muscularis\": \"#CDAA7D\",\n",
    "    \"Lung\": \"#9ACD32\",\n",
    "    \"Muscle_Skeletal\": \"#7A67EE\",\n",
    "    \"Nerve_Tibial\": \"#FFD700\",\n",
    "    \"Skin_Not_Sun_Exposed_Suprapubic\": \"#3A5FCD\",\n",
    "    \"Skin_Sun_Exposed_Lower_leg\": \"#1E90FF\",\n",
    "    \"Thyroid\": \"#008B45\",\n",
    "    \"Whole_Blood\": \"#FF00FF\"\n",
    "}\n",
    "\n",
    "gtex_tissue_abbrev = ['ADPSBQ',\n",
    "                      'ADPVSC',\n",
    "                      'ARTTBL',\n",
    "                      'CELLS',\n",
    "                      'ESPMCS',\n",
    "                      'ESPMSL',\n",
    "                      'LUNG',\n",
    "                      'MSCLSK',\n",
    "                      'NERVET',\n",
    "                      'SKINNS',\n",
    "                      'SKINS',\n",
    "                      'THYROID',\n",
    "                      'WHLBLD']\n",
    "\n",
    "gtex_tissue_pal = sns.color_palette(list(gtex_tissue_dict.values()))\n",
    "gtex_tissue_pal_df = pd.DataFrame(pd.Series(gtex_tissue_dict), columns=['hex']).reset_index(names=['tissue_id'])\n",
    "gtex_tissue_pal\n",
    "\n",
    "# %config InlineBackend.figure_formats = ['svg']\n",
    "# %matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main figure panels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in clusters\n",
    "clusters = load_across_tissues(config, load_clusters_annotated)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### number clusters by tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get total number of clusters and genes in clusters\n",
    "cluster_counts = clusters.groupby('tissue_id').agg({'N_genes':sum, 'Transcripts':'nunique'})\n",
    "tissue_ids = load_tissue_ids(config)\n",
    "\n",
    "# plot showing number of clusters in each tissue\n",
    "fig, ax = plt.subplots(figsize=(2.5,3.5))\n",
    "ax = sns.barplot(cluster_counts,hue_order=tissue_ids, x='Transcripts', hue='tissue_id', \n",
    "                 y='tissue_id', palette=gtex_tissue_pal, ax=ax, width=1, edgecolor='k')\n",
    "ax.set_xlim((0, ax.get_xlim()[1]+100))\n",
    "ax.set_xticks([0,500,1000])\n",
    "ax.set_xlabel('Clusters', fontsize=12)\n",
    "ax.set_ylabel('')\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "\n",
    "# add numbers to plot\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=2)\n",
    "\n",
    "# Replace underscores with tisue abbrevations\n",
    "ax.set_yticklabels(gtex_tissue_abbrev)\n",
    "\n",
    "plt.tight_layout() \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cluster size by correlation sign\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters['corr_type'] = np.where(clusters['Mean_neg_cor'].isna(), 'positive only', np.where(clusters['Mean_pos_cor'].isna(), 'negative only', 'mixed'))\n",
    "clusters['N_genes_clip'] = clusters['N_genes'].clip(0, 20)\n",
    "\n",
    "# histogram not broken up by cluster size\n",
    "fig, ax = plt.subplots(figsize=(3.5,3.5))\n",
    "\n",
    "sns.histplot(clusters, y='N_genes_clip', ax=ax, alpha=.8, discrete=True, hue='corr_type', multiple='dodge', palette=['grey', '#c4553aff', '#3e8093ff'], shrink=.8)\n",
    "ax.set_xlabel('Clusters', fontsize=12)\n",
    "ax.set_ylabel('Number of genes in cluster', fontsize=12)\n",
    "ax.set_ylim([1, 21])\n",
    "ax.set_yticks(np.arange(2, 21)[::2])\n",
    "ax.set_yticklabels([*np.arange(2, 20)[::2], '20+'], fontsize=10)\n",
    "ax.set_xscale('log')\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "\n",
    "# customize the legend\n",
    "legend = ax.get_legend()\n",
    "legend.get_frame().set_visible(False)\n",
    "legend.set_title(\"Gene corelations in cluster:\")\n",
    "\n",
    "plt.tight_layout() \n",
    "plt.show()\n",
    "\n",
    "clusters['N_genes'].clip(0, 4).reset_index().groupby('N_genes').agg({'index':'nunique'})/len(clusters) * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### example chr from fibroblast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_tissue_id = 'Cells_Cultured_fibroblasts'\n",
    "chr_num = 7\n",
    "\n",
    "# load in covatiates\n",
    "covariates_df = pd.read_csv('{}/{}/{}.v8.covariates.txt'.format(prefix, config['covariates_dir'], my_tissue_id), sep='\\t', index_col=0).T\n",
    "base_covariates = covariates_df[['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'pcr', 'platform', 'sex']]\n",
    "peer_covariates = covariates_df[[f'InferredCov{i}' for i in range(1,61,1)]]\n",
    "\n",
    "# load in expression\n",
    "expression_df = load_expression(config, my_tissue_id)\n",
    "expression_df = expression_df.set_index('gene_id')[covariates_df.index]\n",
    "\n",
    "\n",
    "# load in the gene information (for gene-gene distances)\n",
    "gid_gencode, full_gencode = load_gencode()\n",
    "gid_gencode_tissue = gid_gencode.loc[expression_df.index]\n",
    "# get the genes to plot\n",
    "selected_chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{chr_num}'].index.values\n",
    "selected_chr_gene_names = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{chr_num}']['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to residualize expression based on two covariate dfs\n",
    "def two_part_residualize(expression_df, base_covariates_df, additional_covariates_df):\n",
    "    full_covariates = pd.concat([base_covariates_df, additional_covariates_df], axis=1)\n",
    "    # residulize the expression \n",
    "    residal_exp = calculate_residual(expression_df[full_covariates.index], full_covariates, center=True)\n",
    "    residal_exp = pd.DataFrame(residal_exp, columns=full_covariates.index, index=expression_df.index)\n",
    "    return residal_exp\n",
    "\n",
    "\n",
    "# corr with all 60 peers removed\n",
    "peer_residual_exp = two_part_residualize(expression_df, base_covariates, peer_covariates)\n",
    "corrected_corr, corrected_pvalue = spearmanr(peer_residual_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "\n",
    "# plot a zoomed in part\n",
    "\n",
    "# for chr 17 skin sun exposed\n",
    "# start_idx = 417\n",
    "# end_idx = 554\n",
    "start_idx = 0\n",
    "end_idx = len(corrected_corr)\n",
    "\n",
    "corrected_corr = corrected_corr[start_idx:end_idx, start_idx:end_idx]\n",
    "corrected_pvalue = corrected_pvalue[start_idx:end_idx, start_idx:end_idx]\n",
    "\n",
    "# calculate total number of pairs considered for bonferroni correction\n",
    "total_pairs = 0\n",
    "max_cluster_size=50\n",
    "for i in np.arange(1,23,1):\n",
    "    chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{i}'].index.values\n",
    "    upper_corner_idxs = np.triu(np.ones(len(chr_gene_ids)), k=1)\n",
    "    excluded_cluster_size_idxs = np.triu(np.ones(len(chr_gene_ids)), k=max_cluster_size)\n",
    "    total_pairs += upper_corner_idxs.sum()  - excluded_cluster_size_idxs.sum()\n",
    "    \n",
    "# a version with non-signifigant correlations blanked out\n",
    "corrected_corr_masked = corrected_corr.copy()\n",
    "corrected_corr_masked[corrected_pvalue >= (.05/total_pairs)] = 0\n",
    "corrected_corr_masked = pd.DataFrame(corrected_corr_masked, index=selected_chr_gene_names[start_idx:end_idx], columns=selected_chr_gene_names[start_idx:end_idx])\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "sns.heatmap(corrected_corr_masked, mask=np.triu(np.ones_like(corrected_corr), k=1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "\n",
    "# add in clusters\n",
    "for idx, cluster in clusters[(clusters['Chromosome']==chr_num)&(clusters['tissue_id']==my_tissue_id)].iterrows():\n",
    "    cluster_idxs = np.concatenate([np.where(selected_chr_gene_ids[start_idx:end_idx]==cluster['Transcripts'].split(',')[i])[0] for i in range(cluster['N_genes'])])\n",
    "    if len(cluster_idxs) > 0:\n",
    "        # Calculate rectangle properties\n",
    "        x = cluster_idxs[0]\n",
    "        y = cluster_idxs[0]\n",
    "        width = np.ptp(cluster_idxs) + 1\n",
    "        height = np.ptp(cluster_idxs) + 1\n",
    "        ax.plot([x, x + width], [y + height, y + height], color='k', linewidth=1)\n",
    "        ax.plot([x, x], [y, y + height], color='k', linewidth=1)\n",
    "\n",
    "        # ax.add_patch(mpatches.Rectangle((cluster_idxs[0], cluster_idxs[0]), \n",
    "        #                             width=np.ptp(cluster_idxs)+1, height=np.ptp(cluster_idxs)+1,\n",
    "        #                             linewidth=1, edgecolor='k', facecolor='none'))\n",
    "#ax.spines[['bottom', 'left']].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### example chr from skin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in expression and covariates\n",
    "my_tissue_id = 'Skin_Sun_Exposed_Lower_leg'\n",
    "chr_num = 17\n",
    "\n",
    "\n",
    "# load in covatiates\n",
    "covariates_df = pd.read_csv('{}/{}/{}.v8.covariates.txt'.format(prefix, config['covariates_dir'], my_tissue_id), sep='\\t', index_col=0).T\n",
    "base_covariates = covariates_df[['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'pcr', 'platform', 'sex']]\n",
    "peer_covariates = covariates_df[[f'InferredCov{i}' for i in range(1,61,1)]]\n",
    "\n",
    "# load in expression\n",
    "expression_df = load_expression(config, my_tissue_id)\n",
    "expression_df = expression_df.set_index('gene_id')[covariates_df.index]\n",
    "\n",
    "\n",
    "# load in the gene information (for gene-gene distances)\n",
    "gid_gencode, full_gencode = load_gencode()\n",
    "gid_gencode_tissue = gid_gencode.loc[expression_df.index]\n",
    "# get the genes to plot\n",
    "selected_chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{chr_num}'].index.values\n",
    "selected_chr_gene_names = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{chr_num}']['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to residualize expression based on two covariate dfs\n",
    "def two_part_residualize(expression_df, base_covariates_df, additional_covariates_df):\n",
    "    full_covariates = pd.concat([base_covariates_df, additional_covariates_df], axis=1)\n",
    "    # residulize the expression \n",
    "    residal_exp = calculate_residual(expression_df[full_covariates.index], full_covariates, center=True)\n",
    "    residal_exp = pd.DataFrame(residal_exp, columns=full_covariates.index, index=expression_df.index)\n",
    "    return residal_exp\n",
    "\n",
    "\n",
    "# corr with all 60 peers removed\n",
    "peer_residual_exp = two_part_residualize(expression_df, base_covariates, peer_covariates)\n",
    "corrected_corr, corrected_pvalue = spearmanr(peer_residual_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "\n",
    "# plot a zoomed in part\n",
    "\n",
    "# for chr 17 skin sun exposed\n",
    "# start_idx = 417\n",
    "# end_idx = 554\n",
    "start_idx = 0\n",
    "end_idx = len(corrected_corr)\n",
    "\n",
    "corrected_corr = corrected_corr[start_idx:end_idx, start_idx:end_idx]\n",
    "corrected_pvalue = corrected_pvalue[start_idx:end_idx, start_idx:end_idx]\n",
    "\n",
    "# calculate total number of pairs considered for bonferroni correction\n",
    "total_pairs = 0\n",
    "max_cluster_size=50\n",
    "for i in np.arange(1,23,1):\n",
    "    chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{i}'].index.values\n",
    "    upper_corner_idxs = np.triu(np.ones(len(chr_gene_ids)), k=1)\n",
    "    excluded_cluster_size_idxs = np.triu(np.ones(len(chr_gene_ids)), k=max_cluster_size)\n",
    "    total_pairs += upper_corner_idxs.sum()  - excluded_cluster_size_idxs.sum()\n",
    "    \n",
    "# a version with non-signifigant correlations blanked out\n",
    "corrected_corr_masked = corrected_corr.copy()\n",
    "corrected_corr_masked[corrected_pvalue >= (.05/total_pairs)] = 0\n",
    "corrected_corr_masked = pd.DataFrame(corrected_corr_masked, index=selected_chr_gene_names[start_idx:end_idx], columns=selected_chr_gene_names[start_idx:end_idx])\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "sns.heatmap(corrected_corr_masked, mask=np.triu(np.ones_like(corrected_corr), k=1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "\n",
    "# add in clusters\n",
    "for idx, cluster in clusters[(clusters['Chromosome']==chr_num)&(clusters['tissue_id']==my_tissue_id)].iterrows():\n",
    "    cluster_idxs = np.concatenate([np.where(selected_chr_gene_ids[start_idx:end_idx]==cluster['Transcripts'].split(',')[i])[0] for i in range(cluster['N_genes'])])\n",
    "    if len(cluster_idxs) > 0:\n",
    "        # Calculate rectangle properties\n",
    "        x = cluster_idxs[0]\n",
    "        y = cluster_idxs[0]\n",
    "        width = np.ptp(cluster_idxs) + 1\n",
    "        height = np.ptp(cluster_idxs) + 1\n",
    "        ax.plot([x, x + width], [y + height, y + height], color='k', linewidth=1)\n",
    "        ax.plot([x, x], [y, y + height], color='k', linewidth=1)\n",
    "\n",
    "        # ax.add_patch(mpatches.Rectangle((cluster_idxs[0], cluster_idxs[0]), \n",
    "        #                             width=np.ptp(cluster_idxs)+1, height=np.ptp(cluster_idxs)+1,\n",
    "        #                             linewidth=1, edgecolor='k', facecolor='none'))\n",
    "#ax.spines[['bottom', 'left']].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pulled out example chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cluster(example_cluster, cbar=False):\n",
    "    # gene gene ids sorted by tss start\n",
    "    sorted_cluster = gid_gencode.loc[example_cluster['cluster_id'].split('_')].sort_values('tss_start')\n",
    "    sorted_gene_ids = sorted_cluster.index.values\n",
    "    # get normed expression \n",
    "    expression = load_cluster_expression(config, example_cluster['Tissue'])\n",
    "    cluster_expression = expression[(expression['cluster_id']==example_cluster['cluster_id'])].set_index('egene_id').loc[sorted_gene_ids]\n",
    "    # get correlation of expression\n",
    "    sample_ids = cluster_expression.columns[cluster_expression.columns.str.contains('GTEX')]\n",
    "    cluster_corr, cluster_pvalue = spearmanr(cluster_expression[sample_ids], axis=1)\n",
    "    # make df wtih readable gene names\n",
    "    cluster_corr = pd.DataFrame(cluster_corr, index=sorted_cluster['gene_name'].values, columns=sorted_cluster['gene_name'].values)\n",
    "\n",
    "    # make corr plot\n",
    "    \n",
    "    fig_size = int(len(cluster_corr)/2)\n",
    "    print(fig_size)\n",
    "    fig, ax = plt.subplots(figsize=(fig_size, fig_size))\n",
    "    sns.heatmap(cluster_corr, mask=np.triu(np.ones_like(cluster_corr), k=1),\n",
    "                cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, \n",
    "                cbar_kws={'label':'Spearman Corrleation'}, \n",
    "                xticklabels=True, yticklabels=True, cbar=cbar)\n",
    "    ax.tick_params(\"y\", rotation=0) \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cluster(clusters.iloc[3132], cbar=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cluster(clusters.iloc[9757])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cluster enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get just the ones with matched enahncer\n",
    "enhancer_tissue_ids = ['Artery_Tibial', 'Cells_Cultured_fibroblasts', 'Lung', 'Muscle_Skeletal', 'Skin_Not_Sun_Exposed_Suprapubic', 'Skin_Sun_Exposed_Lower_leg', 'Thyroid']\n",
    "combined_clusters = load_across_tissues(config, load_clusters_annotated, tissue_ids = enhancer_tissue_ids)\n",
    "def load_all_multigene_null(config, tissue_id):\n",
    "    return pd.concat([load_null_clusters_annotated(config, tissue_id, num_genes) for num_genes in [2,3,4,5]])\n",
    "combined_multigene_nulls =  load_across_tissues(config, load_all_multigene_null, tissue_ids = enhancer_tissue_ids)\n",
    "\n",
    "# exclude highly cross mappable genes and add columns\n",
    "combined_clusters = combined_clusters[~combined_clusters['has_cross_map']]\n",
    "combined_clusters['has_multiple_abc_genes'] = combined_clusters['num_abc_genes'] > 1\n",
    "combined_clusters['log_size'] = np.log10(combined_clusters['cluster_tss_size'])\n",
    "combined_multigene_nulls = combined_multigene_nulls[~combined_multigene_nulls['has_cross_map']]\n",
    "combined_multigene_nulls['has_multiple_abc_genes'] = combined_multigene_nulls['num_abc_genes'] > 1\n",
    "combined_multigene_nulls['log_size'] = np.log10(combined_multigene_nulls['cluster_tss_size'])\n",
    "\n",
    "\n",
    "# make nulls  \n",
    "\n",
    "# make one without distance matching\n",
    "multitissue_abc_largerclusters_df = pd.concat([combined_clusters, combined_multigene_nulls], keys=['cluster', 'null'], names=['type', 'idx']).reset_index()\n",
    "multitissue_abc_largerclusters_df['is_cluster'] = multitissue_abc_largerclusters_df['type']=='cluster'\n",
    "\n",
    "# make one with distance matching\n",
    "multitissue_abc_resamp_largerclusters_nulls = []\n",
    "for num_genes in combined_multigene_nulls['N_genes'].unique():\n",
    "    cluster_num_genes = combined_clusters[(combined_clusters['N_genes']==num_genes)]\n",
    "    # don't bother to match if too few clusters of this size, just discard null\n",
    "    if len(cluster_num_genes) > 2:\n",
    "        null_num_genes = combined_multigene_nulls[(combined_multigene_nulls['N_genes']==num_genes)]\n",
    "        multitissue_abc_resamp_largerclusters_nulls.append(get_resamp_null_cluster(null_num_genes, cluster_num_genes, number_null=5000*len(enhancer_tissue_ids), plot=False))\n",
    "\n",
    "multitissue_abc_resamp_largerclusters_df = pd.concat([combined_clusters, pd.concat(multitissue_abc_resamp_largerclusters_nulls)], keys=['cluster', 'null'], names=['type', 'idx']).reset_index()\n",
    "multitissue_abc_resamp_largerclusters_df['is_cluster'] = multitissue_abc_resamp_largerclusters_df['type']=='cluster'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multitissue_abc_largerclusters_df['corr_type'] = np.where(multitissue_abc_largerclusters_df['Mean_neg_cor'].isna(), 'positive_only', np.where(multitissue_abc_largerclusters_df['Mean_pos_cor'].isna(), 'negative_only', 'mixed'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get odds df for positive negative, and mixed corr clusters\n",
    "\n",
    "multitissue_abc_largerclusters_df['corr_type'] = np.where(multitissue_abc_largerclusters_df['Mean_neg_cor'].isna(), 'positive_only', np.where(multitissue_abc_largerclusters_df['Mean_pos_cor'].isna(), 'negative_only', 'mixed'))\n",
    "\n",
    "noresamp_cols = ['has_bidirectional_promoter']\n",
    "noresamp_odds_positive = get_odds_df(multitissue_abc_largerclusters_df[((multitissue_abc_largerclusters_df['type']=='cluster')&(multitissue_abc_largerclusters_df['corr_type']=='positive_only')) | (multitissue_abc_largerclusters_df['type']=='null')], column_list = noresamp_cols, correct_on=True, correct_on_column='N_genes', label_col='is_cluster')\n",
    "noresamp_odds_all = get_odds_df(multitissue_abc_largerclusters_df[((multitissue_abc_largerclusters_df['type']=='cluster')&(multitissue_abc_largerclusters_df['corr_type']=='mixed')) | (multitissue_abc_largerclusters_df['type']=='null')], column_list = noresamp_cols, correct_on=True, correct_on_column='N_genes', label_col='is_cluster')\n",
    "noresamp_odds_negative = get_odds_df(multitissue_abc_largerclusters_df[((multitissue_abc_largerclusters_df['type']=='cluster')&(multitissue_abc_largerclusters_df['corr_type']=='negative_only')) | (multitissue_abc_largerclusters_df['type']=='null')], column_list = noresamp_cols, correct_on=True, correct_on_column='N_genes', label_col='is_cluster')\n",
    "\n",
    "\n",
    "multitissue_abc_resamp_largerclusters_df['corr_type'] = np.where(multitissue_abc_resamp_largerclusters_df['Mean_neg_cor'].isna(), 'positive_only', np.where(multitissue_abc_resamp_largerclusters_df['Mean_pos_cor'].isna(), 'negative_only', 'mixed'))\n",
    "\n",
    "resamp_column_list = ['has_tads_tss', \n",
    "                      'has_paralog', \n",
    "                      'has_shared_go_any',\n",
    "                      'has_ctcf_peak',\n",
    "                      'has_shared_enhancer']\n",
    "\n",
    "bool_filter_list = ['has_shared_enhancer']\n",
    "\n",
    "resamp_odds_positive = get_odds_df(multitissue_abc_resamp_largerclusters_df[((multitissue_abc_resamp_largerclusters_df['type']=='cluster')&(multitissue_abc_resamp_largerclusters_df['corr_type']=='positive_only')) | (multitissue_abc_resamp_largerclusters_df['type']=='null')], column_list=resamp_column_list, filter_list=bool_filter_list, correct_on=True, correct_on_column='N_genes', label_col='is_cluster') \n",
    "resamp_odds_all = get_odds_df(multitissue_abc_resamp_largerclusters_df[((multitissue_abc_resamp_largerclusters_df['type']=='cluster')&(multitissue_abc_resamp_largerclusters_df['corr_type']=='mixed')) | (multitissue_abc_resamp_largerclusters_df['type']=='null')], column_list=resamp_column_list, filter_list=bool_filter_list, correct_on=True, correct_on_column='N_genes', label_col='is_cluster')\n",
    "resamp_odds_negative = get_odds_df(multitissue_abc_resamp_largerclusters_df[((multitissue_abc_resamp_largerclusters_df['type']=='cluster')&(multitissue_abc_resamp_largerclusters_df['corr_type']=='negative_only')) | (multitissue_abc_resamp_largerclusters_df['type']=='null')], column_list=resamp_column_list, filter_list=bool_filter_list, correct_on=True, correct_on_column='N_genes', label_col='is_cluster')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combined, resampled and nonresampled columns\n",
    "col_order = ['has_tads_tss', 'has_ctcf_peak', 'has_shared_enhancer', 'has_bidirectional_promoter', 'has_shared_go_any', 'has_paralog'][::-1]\n",
    "odds_negative = pd.concat([noresamp_odds_negative, resamp_odds_negative]).loc[col_order] \n",
    "odds_all = pd.concat([noresamp_odds_all, resamp_odds_all]).loc[col_order] \n",
    "odds_positive = pd.concat([noresamp_odds_positive, resamp_odds_positive]).loc[col_order] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3.5, 3.5))\n",
    "\n",
    "# make the plot\n",
    "make_log_odds_plot_multiple([odds_negative, odds_positive], \n",
    "                             labels=['negative', 'positive',], \n",
    "                             colors = sns.color_palette('blend:#67AFD2,k,#B83A4B', n_colors=2), \n",
    "                             add_annotations=False, offset=.3, ax=ax)\n",
    "\n",
    "ax.set_xlabel(r'$log_{10}$(Fold enrichment correlated clusters)', fontsize=10)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.set_yticklabels([])\n",
    "ax.set_xlim([.4, 11])\n",
    "ax.set_xticks([.5, 1, 5, 10], labels=[.5, 1, 5, 10])\n",
    "\n",
    "# customize the legend\n",
    "legend = ax.get_legend()\n",
    "legend.get_frame().set_visible(False)\n",
    "legend.set_title(\"Gene correlaitons in cluster:\")\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1.1, 1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the fraction of True values for each column in col_order\n",
    "fraction_true = combined_clusters[col_order].mean()\n",
    "\n",
    "# Create a vertical bar graph\n",
    "plt.figure(figsize=(2.5, 3.5))\n",
    "ax = sns.barplot(y=fraction_true.index, x=fraction_true.values, color='grey')\n",
    "ax.bar_label(ax.containers[0], fmt='%.2f')\n",
    "ax.set_ylabel(\" \")\n",
    "ax.set_xlabel(\"Proportion of clusters\", fontsize=12)\n",
    "ax.set_xlim([0,1])\n",
    "ax.set_yticklabels(np.array(['paralogs',  'shared GO',  'promoter', 'Enhancer',  'CTCF',  'crosses TAD'][::-1]), fontsize=12)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a) chromosome correlaiton before and after peer correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate total number of pairs considered for bonferroni correction\n",
    "total_pairs = 0\n",
    "max_cluster_size=50\n",
    "for i in np.arange(1,23,1):\n",
    "    chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{i}'].index.values\n",
    "    upper_corner_idxs = np.triu(np.ones(len(chr_gene_ids)), k=1)\n",
    "    excluded_cluster_size_idxs = np.triu(np.ones(len(chr_gene_ids)), k=max_cluster_size)\n",
    "    total_pairs += upper_corner_idxs.sum()  - excluded_cluster_size_idxs.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inital corr with no peer factors removed\n",
    "residal_exp = calculate_residual(expression_df[base_covariates.index], base_covariates, center=True)\n",
    "residal_exp = pd.DataFrame(residal_exp, columns=base_covariates.index, index=expression_df.index)\n",
    "inital_corr, initial_pvalue = spearmanr(residal_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "\n",
    "# a version with non-signifigant correlations blanked out\n",
    "inital_corr_masked = inital_corr.copy()\n",
    "inital_corr_masked[initial_pvalue >= (.05/total_pairs)] = 0\n",
    "inital_corr_masked = pd.DataFrame(inital_corr_masked, index=selected_chr_gene_names, columns=selected_chr_gene_names)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "sns.heatmap(inital_corr_masked, mask=np.triu(np.ones_like(inital_corr), k=1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "ax.spines[['bottom', 'left']].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to residualize expression based on two covariate dfs\n",
    "def two_part_residualize(expression_df, base_covariates_df, additional_covariates_df):\n",
    "    full_covariates = pd.concat([base_covariates_df, additional_covariates_df], axis=1)\n",
    "    # residulize the expression \n",
    "    residal_exp = calculate_residual(expression_df[full_covariates.index], full_covariates, center=True)\n",
    "    residal_exp = pd.DataFrame(residal_exp, columns=full_covariates.index, index=expression_df.index)\n",
    "    return residal_exp\n",
    "\n",
    "\n",
    "# corr with all 60 peers removed\n",
    "peer_residual_exp = two_part_residualize(expression_df, base_covariates, peer_covariates)\n",
    "corrected_corr, corrected_pvalue = spearmanr(peer_residual_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "\n",
    "# plot a zoomed in part\n",
    "\n",
    "# for chr 17 skin sun exposed\n",
    "# start_idx = 417\n",
    "# end_idx = 554\n",
    "start_idx = 0\n",
    "end_idx = len(corrected_corr)\n",
    "\n",
    "corrected_corr = corrected_corr[start_idx:end_idx, start_idx:end_idx]\n",
    "corrected_pvalue = corrected_pvalue[start_idx:end_idx, start_idx:end_idx]\n",
    "\n",
    "\n",
    "# a version with non-signifigant correlations blanked out\n",
    "corrected_corr_masked = corrected_corr.copy()\n",
    "corrected_corr_masked[corrected_pvalue >= (.05/total_pairs)] = 0\n",
    "corrected_corr_masked = pd.DataFrame(corrected_corr_masked, index=selected_chr_gene_names[start_idx:end_idx], columns=selected_chr_gene_names[start_idx:end_idx])\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "sns.heatmap(corrected_corr_masked, mask=np.triu(np.ones_like(corrected_corr), k=1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "\n",
    "# add in clusters\n",
    "for idx, cluster in clusters[(clusters['Chromosome']==chr_num)&(clusters['tissue_id']==my_tissue_id)].iterrows():\n",
    "    cluster_idxs = np.concatenate([np.where(selected_chr_gene_ids[start_idx:end_idx]==cluster['Transcripts'].split(',')[i])[0] for i in range(cluster['N_genes'])])\n",
    "    if len(cluster_idxs) > 0:\n",
    "        # Calculate rectangle properties\n",
    "        x = cluster_idxs[0]\n",
    "        y = cluster_idxs[0]\n",
    "        width = np.ptp(cluster_idxs) + 1\n",
    "        height = np.ptp(cluster_idxs) + 1\n",
    "        ax.plot([x, x + width], [y + height, y + height], color='k', linewidth=1)\n",
    "        ax.plot([x, x], [y, y + height], color='k', linewidth=1)\n",
    "\n",
    "        # ax.add_patch(mpatches.Rectangle((cluster_idxs[0], cluster_idxs[0]), \n",
    "        #                             width=np.ptp(cluster_idxs)+1, height=np.ptp(cluster_idxs)+1,\n",
    "        #                             linewidth=1, edgecolor='k', facecolor='none'))\n",
    "#ax.spines[['bottom', 'left']].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b) quantification of correlaiton distances before and after peer correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signifigant_distances(gid_gencode_tissue, expression_df, base_covariates, peer_covariates, chr_num, residualization_values  = [0,15,30,45,60]):\n",
    "    selected_chr_gene_ids = gid_gencode_tissue[gid_gencode_tissue['chr'] == f'chr{chr_num}'].index.values\n",
    "    # df with pairswise distances\n",
    "    start_positions = gid_gencode_tissue.loc[selected_chr_gene_ids]['tss_start'].values\n",
    "    dist_matrix = np.abs(start_positions[:, None] - start_positions)\n",
    "\n",
    "    # distance of signifigant corrs as pc/peers residualized away? \n",
    "    signifigant_distances = {}\n",
    "    for i in residualization_values:\n",
    "        # expression residualized with some pcs\n",
    "        sub_exp = two_part_residualize(expression_df, base_covariates, peer_covariates.iloc[:, :i])\n",
    "        pc_residual_corr, pc_residual_pvalue = spearmanr(sub_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "        pc_residual_pvalue[np.triu_indices(len(pc_residual_pvalue))] = 1\n",
    "        signifigant_distances[i] = (dist_matrix[pc_residual_pvalue<(.05/total_pairs)])\n",
    "    return signifigant_distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop through all tissues, takes ~ 5 minutes\n",
    "all_chr_distances = []\n",
    "for my_tissue_id in load_tissue_ids(config):\n",
    "    print(my_tissue_id)\n",
    "    covariates_df = pd.read_csv('{}/{}/{}.v8.covariates.txt'.format(prefix, config['covariates_dir'], my_tissue_id), sep='\\t', index_col=0).T\n",
    "    base_covariates = covariates_df[['PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'pcr', 'platform', 'sex']]\n",
    "    peer_covariates = covariates_df[[f'InferredCov{i}' for i in range(1,61,1)]]\n",
    "    expression_df = load_expression(config, my_tissue_id)\n",
    "    expression_df = expression_df.set_index('gene_id')[covariates_df.index]\n",
    "    gid_gencode_tissue =  gid_gencode.loc[expression_df.index]\n",
    "\n",
    "    # loop through chroms\n",
    "    for i in tqdm(range(1, 23), total=23):\n",
    "        all_chr_distances.append(get_signifigant_distances(gid_gencode_tissue, expression_df, base_covariates, peer_covariates, \n",
    "                                                        i, residualization_values  = [0,60]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine over chroms and tissues\n",
    "combined_chr_distances = {}\n",
    "for i in [0,60]:\n",
    "    combined_chr_distances[i] = np.concatenate([chr_distances[i] for chr_distances in all_chr_distances])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpl.rcParams['figure.dpi']= 1000\n",
    "fig, ax = plt.subplots(figsize=(4,5))\n",
    "sns.violinplot(combined_chr_distances, ax=ax, color='grey', fill=False, cut=0)\n",
    "ax.set_ylabel('Correlated genes\\' distance\\' (bp)', fontsize=12)\n",
    "ax.set_xticklabels(['Unresidualized', 'Residualized'], fontsize=12)\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "plt.show()\n",
    "mpl.rcParams['figure.dpi']= 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c) examples of clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fraction of genes in clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in expression data (so we only compare to pairs of genes also expressed in this tissue)\n",
    "# load in expression data for all tissues \n",
    "# takes about 30 seconds\n",
    "expression={}\n",
    "for tissue_id in tissue_ids: \n",
    "    expression[tissue_id] = load_expression(config, tissue_id)\n",
    "    expression[tissue_id] = expression[tissue_id].sort_values(['#chr', 'start', 'end'])\n",
    "    # we only need the transcript ids from the expression dir \n",
    "    expression[tissue_id].drop(columns=expression[tissue_id].columns[4:], inplace=True)\n",
    "    \n",
    "expression_df = pd.concat(expression, keys=tissue_ids, names=('tissue_id', 'idx')).reset_index('idx', drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_genes_in_clusters=[]\n",
    "for tissue_id in tissue_ids: \n",
    "    all_cluster_transcripts = np.concatenate(clusters[clusters['tissue_id'] == tissue_id]['Transcripts'].str.split(',').values)\n",
    "    frac_genes_in_clusters.append(sum(expression[tissue_id]['gene_id'].isin(all_cluster_transcripts))/len(expression[tissue_id])*100)\n",
    "\n",
    "frac_df = pd.DataFrame({'tissue_id': tissue_ids, 'fraction':frac_genes_in_clusters})\n",
    "print(frac_df['fraction'].mean())\n",
    "print(frac_df['fraction'].min())\n",
    "print(frac_df['fraction'].max())\n",
    "\n",
    "# plot\n",
    "fig, ax = plt.subplots(figsize=(5,5))\n",
    "\n",
    "sns.barplot(pd.DataFrame({'tissue_id': tissue_ids, 'fraction':frac_genes_in_clusters}), x='fraction', y='tissue_id', palette=gtex_tissue_pal, \n",
    "            ax =ax, saturation=1, edgecolor='k', width=1)\n",
    "ax.set_xlabel('Fraction expressed genes in clusters')\n",
    "ax.set_ylabel('')\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "\n",
    "# add numbers to plot\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=2, fmt=f'%.1f')\n",
    "\n",
    "# Replace underscores with tisue abbrevations\n",
    "ax.set_yticklabels(gtex_tissue_abbrev)\n",
    "\n",
    "plt.tight_layout() \n",
    "plt.show()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene orientation/ relative position\n",
    " also the .genes. gtf does not necessaritly have accurate start/end positions for genes due to collapse\n",
    "\n",
    "\n",
    "\n",
    "types of genes (heirarchy so top ones are annotated even if bottom ones are true too)\n",
    "* shared_promoter (promoter defined 2000 bp upsteam of transcript start to 200 bp downstream of transcript start) same or opp orientiaton \n",
    "* convergent/divergent/tandem (no promoter overlap, with or without _overlapping tag for convergent and tandem)\n",
    "* any transcirpt overlap? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_v26=pr.read_gtf('/home/klawren/oak/pcqtls/data/references/gencode.v26.annotation.gtf.gz').as_df()\n",
    "\n",
    "# this is all done at the transcript level\n",
    "gencode_v26 = gencode_v26[gencode_v26['Feature'] == 'transcript'].set_index('gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_v26['tss_start'] =  np.where(gencode_v26['Strand'] == '+', gencode_v26['Start'], gencode_v26['End'])\n",
    "gencode_v26['promoter_start'] = np.where(gencode_v26['Strand'] == '+', gencode_v26['tss_start'] - 1000, gencode_v26['tss_start'] - 100)\n",
    "gencode_v26['promoter_end'] = np.where(gencode_v26['Strand'] == '+', gencode_v26['tss_start'] + 100, gencode_v26['tss_start'] + 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make clusters have a row per gene comparision that we are doing\n",
    "clusters['gene_id_1'] = clusters['Transcripts'].str.split(',')\n",
    "clusters['gene_id_2'] = clusters['Transcripts'].str.split(',')\n",
    "clusters_explode = clusters.explode('gene_id_1').explode('gene_id_2')\n",
    "clusters_explode = clusters_explode[~(clusters_explode['gene_id_1'] == clusters_explode['gene_id_2'])]\n",
    "clusters_explode['gene_id_pair'] = clusters_explode['gene_id_1'] + '_' + clusters_explode['gene_id_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = clusters.iloc[19]\n",
    "gene_1 = row['Transcripts'].split(',')[0]\n",
    "gene_2 = row['Transcripts'].split(',')[1]\n",
    "gene_1_gencode = gencode_v26.loc[gene_1]\n",
    "gene_2_gencode = gencode_v26.loc[gene_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_1_gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_2_gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_1_gencode['tss_start'].max() > gene_2_gencode['tss_start'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_1_gencode['Strand'][0]=='+', gene_1_gencode['tss_start'].max() > gene_2_gencode['tss_start'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_relative_position(gene_1_gencode, gene_2_gencode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_relative_position(gene_1_gencode, gene_2_gencode):\n",
    "    if (gene_1_gencode['Strand'][0] == gene_2_gencode['Strand'][0]) > 0:\n",
    "        orientaiton = 'same'\n",
    "    else:\n",
    "        orientaiton = 'opp'\n",
    "\n",
    "    # check for shared promoter\n",
    "    overlapping_promoter=False\n",
    "    overlapping_transcript=False\n",
    "\n",
    "    try:\n",
    "        for idx1, gene_1_row in gene_1_gencode.iterrows():\n",
    "            if (((gene_1_row['promoter_start'] < gene_2_gencode['promoter_end']) & (gene_1_row['promoter_start'] > gene_2_gencode['promoter_start'])).any() |\n",
    "                ((gene_1_row['promoter_end'] < gene_2_gencode['promoter_end']) & (gene_1_row['promoter_end'] > gene_2_gencode['promoter_start'])).any()):\n",
    "                overlapping_promoter=True\n",
    "            if (((gene_1_row['Start'] < gene_2_gencode['End']) & (gene_1_row['Start'] > gene_2_gencode['Start'])).any() |\n",
    "                ((gene_1_row['End'] < gene_2_gencode['End']) & (gene_1_row['End'] > gene_2_gencode['Start'])).any()):\n",
    "                overlapping_transcript=True\n",
    "    except AttributeError:\n",
    "        gene_1_row = gene_1_gencode\n",
    "        if (((gene_1_row['promoter_start'] < gene_2_gencode['promoter_end']) & (gene_1_row['promoter_start'] > gene_2_gencode['promoter_start'])).any() |\n",
    "        ((gene_1_row['promoter_end'] < gene_2_gencode['promoter_end']) & (gene_1_row['promoter_end'] > gene_2_gencode['promoter_start'])).any()):\n",
    "            overlapping_promoter=True\n",
    "        if (((gene_1_row['Start'] < gene_2_gencode['End']) & (gene_1_row['Start'] > gene_2_gencode['Start'])).any() |\n",
    "            ((gene_1_row['End'] < gene_2_gencode['End']) & (gene_1_row['End'] > gene_2_gencode['Start'])).any()):\n",
    "            overlapping_transcript=True\n",
    "\n",
    "\n",
    "    print(orientaiton)\n",
    "    print(overlapping_promoter)\n",
    "    print(overlapping_transcript)\n",
    "\n",
    "    if overlapping_promoter:\n",
    "        if orientaiton=='same':\n",
    "            return 'shared_promoter'\n",
    "        elif orientaiton=='opp':\n",
    "            return 'bidirectional_promoter'\n",
    "    elif overlapping_transcript:\n",
    "        if orientaiton=='same':\n",
    "            return 'tandem_overlapping'\n",
    "        elif orientaiton=='opp':\n",
    "            return 'convergent_overlapping'\n",
    "    else:\n",
    "        if orientaiton=='same':\n",
    "            return 'tandem'\n",
    "        elif gene_1_gencode['Strand'][0]=='+':\n",
    "            if gene_1_gencode['tss_start'].max() > gene_2_gencode['tss_start'].max():\n",
    "                return 'divergent'\n",
    "            else:\n",
    "                return 'convergent'\n",
    "        else:\n",
    "            if gene_1_gencode['tss_start'].max() > gene_2_gencode['tss_start'].max():\n",
    "                return 'convergent'\n",
    "            else:\n",
    "                return 'divergent'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# start with most lenient annotations then back fill\n",
    "\n",
    "# convergent non overlapping\n",
    "clusters['']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### correlaitons before and after correciton with all showing, not just signifigant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap corr without non-signifignat blanked out \n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "sns.heatmap(inital_corr, mask=np.triu(np.ones_like(inital_corr), k=1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "ax.spines['bottom'].set_visible(True)\n",
    "ax.spines['left'].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()\n",
    "\n",
    "# for supp: version without non-signifignat masked\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "sns.heatmap(corrected_corr, mask=np.tril(np.ones_like(corrected_corr), k=-1), cmap=corr_cmap, vmin=-1, vmax=1, ax=ax, cbar_kws={'label':'Spearman Corrleation', 'pad':0}, xticklabels=False, yticklabels=False, cbar=False)\n",
    "ax.spines['top'].set_visible(True)\n",
    "ax.spines['right'].set_visible(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCs vs PEERs correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get pc covariantes (these will be cacluated after the genotype pc and other base covars are residualized)\n",
    "\n",
    "base_residual_expression = calculate_residual(expression_df[base_covariates.index], base_covariates, center=True)\n",
    "base_residual_expression = pd.DataFrame(base_residual_expression, columns=base_covariates.index, index=expression_df.index)\n",
    "\n",
    "# get global pcs \n",
    "pca = PCA()\n",
    "global_pc_values = pca.fit_transform(base_residual_expression.T)\n",
    "pc_covariates_df = pd.DataFrame(global_pc_values, index=covariates_df.index, columns=[f'PC{i}' for i in range(global_pc_values.shape[0])])\n",
    "\n",
    "# compare pc values to peer values\n",
    "pc_peer_corr=np.zeros((60,60))\n",
    "for i in range(60):\n",
    "    for j in range(60):\n",
    "        pc_peer_corr[i,j] = spearmanr(peer_covariates.T.iloc[i], pc_covariates_df.T.iloc[j]).statistic\n",
    "\n",
    "blue_cmap = LinearSegmentedColormap.from_list('mycmap', [(0, 'white'), (1, '#4298B5')])\n",
    "\n",
    "ax = sns.heatmap(np.square(pc_peer_corr), vmin=0, vmax=1, cmap=blue_cmap, cbar_kws={'label':'Spearman $r^2$'})\n",
    "plt.xlabel('PEER factor index')\n",
    "plt.ylabel('PCA factor index')\n",
    "ax.set_xticks(ticks = [0, 14, 29, 44, 59], labels=[1, 15, 30, 45, 60])\n",
    "ax.set_yticks(ticks = [0, 14, 29, 44, 59], labels=[1, 15, 30, 45, 60])\n",
    "ax.invert_yaxis()\n",
    "\n",
    "ax.spines[['top', 'bottom', 'left', 'right']].set_visible(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCs vs PEERs long and short range correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_frac_sig_at_decile(residual_pvalue, decile_df, short_decile_cutoff=1, long_decile_cutoff=8):\n",
    "      fraction_close_sig = []\n",
    "      fraction_far_sig = []\n",
    "\n",
    "\n",
    "      # sig corrs by decile\n",
    "      close_pvals = residual_pvalue[decile_df<short_decile_cutoff]\n",
    "      fraction_close_sig = sum(close_pvals < (.05/total_pairs))/len(close_pvals)\n",
    "\n",
    "      # signifigant corrs off the diagonal\n",
    "      #far_pvals = peer_residual_pvalue[np.triu_indices(len(peer_residual_pvalue), k=3)]\n",
    "\n",
    "      # sig corrs by decile\n",
    "      far_pvals = residual_pvalue[decile_df>long_decile_cutoff]\n",
    "      fraction_far_sig = ((far_pvals < (.05/total_pairs))/len(far_pvals)).sum()\n",
    "\n",
    "      return fraction_close_sig, fraction_far_sig\n",
    "\n",
    "def compare_peer_pc(expression_df, base_covariates_df, pc_covariates_df, peer_covariates_df, decile_df, use_mask=False, mask=None):\n",
    "\n",
    "   fraction_close_sig_pc = []\n",
    "   fraction_far_sig_pc = []\n",
    "\n",
    "   fraction_close_sig_peer = []\n",
    "   fraction_far_sig_peer = []\n",
    "\n",
    "   for i in range(60):\n",
    "      # expression residualized with some pcs\n",
    "      sub_exp = two_part_residualize(expression_df, base_covariates_df, pc_covariates_df.iloc[:, :i])\n",
    "      pc_residual_corr, pc_residual_pvalue = spearmanr(sub_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "      np.fill_diagonal(pc_residual_pvalue, 1)\n",
    "      if use_mask:\n",
    "         pc_residual_pvalue = np.where(mask, 1, pc_residual_pvalue)\n",
    "\n",
    "      fraction_close_sig, fraction_far_sig = get_frac_sig_at_decile(pc_residual_pvalue, decile_df)\n",
    "\n",
    "      fraction_close_sig_pc.append(fraction_close_sig)\n",
    "      fraction_far_sig_pc.append(fraction_far_sig)\n",
    "\n",
    "      # expression residualized with some peers\n",
    "      sub_exp = two_part_residualize(expression_df, base_covariates_df, peer_covariates_df.iloc[:, :i])\n",
    "      peer_residual_corr, peer_residual_pvalue = spearmanr(sub_exp.loc[selected_chr_gene_ids], axis=1)\n",
    "      np.fill_diagonal(peer_residual_pvalue, 1)\n",
    "\n",
    "      if use_mask:\n",
    "         peer_residual_pvalue = np.where(mask, 1, peer_residual_pvalue)\n",
    "\n",
    "      fraction_close_sig, fraction_far_sig = get_frac_sig_at_decile(peer_residual_pvalue, decile_df)\n",
    "      \n",
    "      fraction_close_sig_peer.append(fraction_close_sig)\n",
    "      fraction_far_sig_peer.append(fraction_far_sig)\n",
    "\n",
    "   if use_mask:\n",
    "      fraction_corr_sig_df = pd.DataFrame({'(mask) PC: short-range correlations':fraction_close_sig_pc, \n",
    "                                     '(mask) PC: long-range correlations':fraction_far_sig_pc, \n",
    "                                     '(mask) PEER: short-range correlations':fraction_close_sig_peer, \n",
    "                                     '(mask) PEER: long-range correlations':fraction_far_sig_peer})\n",
    "\n",
    "   else:\n",
    "      fraction_corr_sig_df = pd.DataFrame({'PC: short-range correlations':fraction_close_sig_pc, \n",
    "                                     'PC: long-range correlations':fraction_far_sig_pc, \n",
    "                                     'PEER: short-range correlations':fraction_close_sig_peer, \n",
    "                                     'PEER: long-range correlations':fraction_far_sig_peer})\n",
    "   return fraction_corr_sig_df \n",
    "\n",
    "# matrix of gene-gene distances\n",
    "start_positions = gid_gencode.loc[selected_chr_gene_ids]['tss_start'].values\n",
    "dist_matrix = np.abs(start_positions[:, None] - start_positions)\n",
    "\n",
    "# Create deciles based on distance \n",
    "deciles = np.percentile(dist_matrix, [10, 20, 30, 40, 50, 60, 70, 80, 90])\n",
    "decile_df = pd.DataFrame(np.digitize(dist_matrix, deciles), index=selected_chr_gene_ids, columns=selected_chr_gene_ids)\n",
    "\n",
    "fraction_corr_sig = compare_peer_pc(expression_df, base_covariates, pc_covariates_df, peer_covariates, decile_df).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fraction_corr_sig_melt = pd.melt(fraction_corr_sig, value_name='frac_corr', id_vars='index')\n",
    "fraction_corr_sig_melt['Factor type'] = fraction_corr_sig_melt['variable'].str.split(':').str[0]\n",
    "#fraction_corr_sig_melt['Gene-gene distance'] = fraction_corr_sig_melt['variable'].str.split(': ').str[1].str.split('-').str[0]\n",
    "fraction_corr_sig_melt['Gene-gene distance'] = np.where(fraction_corr_sig_melt['variable'].str.contains('short'), 'Short (first decile)', 'Long (last decile)')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7,4))\n",
    "sns.scatterplot(fraction_corr_sig_melt, x='index', y='frac_corr', ax=ax, hue='Factor type',\n",
    "                     style='Gene-gene distance', palette=('#4298B5', '#6FA287'), alpha=.8, edgecolor=None)\n",
    "\n",
    "ax.set_xlabel('Number of global factors residualized')\n",
    "ax.set_ylabel('Fraction signifigant correlations')\n",
    "ax.set_xticks(ticks = [0, 14, 29, 44, 59], labels=[1, 15, 30, 45, 60])\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "ax.set_yscale('log')\n",
    "ax.set_yticks(ticks = [.5, .1, .01, .001], labels=[1/2, 1/10, 1/100, 1/1000])\n",
    "\n",
    "plt.gca().get_legend().set_frame_on(False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### shuffled gene locations null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import call_clusters\n",
    "real_cluster_df = load_cluster(config, my_tissue_id)\n",
    "expression_df_full = load_expression(config, my_tissue_id)\n",
    "\n",
    "# cluster density on per chrom basis for the actual clusters\n",
    "cluster_density = []\n",
    "for i in np.arange(1,23,1):\n",
    "    chr_gene_ids = expression_df_full[expression_df_full['#chr'] == f'chr{i}']['gene_id']\n",
    "    cluster_density.append(real_cluster_df[real_cluster_df['Chromosome'] == i]['N_genes'].sum()/len(chr_gene_ids))\n",
    "\n",
    "\n",
    "# randomly resample the dataframe to shuffle the gene locations\n",
    "# reasign the same 'gene_ids' to match the expression df\n",
    "# on all chrs takes ~ 10min\n",
    "final_residual_exp = two_part_residualize(expression_df, base_covariates, peer_covariates)\n",
    "shuffled_expression = final_residual_exp.sample(len(final_residual_exp))\n",
    "shuffled_expression.index = final_residual_exp.index\n",
    "null_cluster_df = call_clusters.get_clusters(expression_df_full, shuffled_expression, my_tissue_id)\n",
    "\n",
    "# cluster density on per chrom basis for the null clusters\n",
    "cluster_density_null = []\n",
    "for i in np.arange(1,23,1):\n",
    "    chr_gene_ids = expression_df_full[expression_df_full['#chr'] == f'chr{i}']['gene_id']\n",
    "    cluster_density_null.append(null_cluster_df[null_cluster_df['Chromosome'] == i]['N_genes'].sum()/len(chr_gene_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine into a plot\n",
    "cluster_density_df = pd.DataFrame({'Real gene locations': cluster_density, 'Shuffled gene locations':cluster_density_null})\n",
    "# plot swarmplot\n",
    "ax = sns.swarmplot(cluster_density_df,color='k', zorder=0, marker=\"$\\circ$\", ec=\"face\", s=10, alpha=.8)\n",
    "# plot boxplot\n",
    "sns.boxplot(cluster_density_df, \n",
    "                 showcaps=True,boxprops={'facecolor':'None'},\n",
    "                 showfliers=False, whiskerprops={'linewidth':1}, ax=ax)\n",
    "ax.set_ylabel('Fraction genes in clusters')\n",
    "ax.spines[['top','right']].set_visible(False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cross mappable clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove cross mappable \n",
    "clusters_annotated = load_across_tissues(config, load_clusters_annotated)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "sns.histplot(clusters_annotated, x='num_cross_map', hue='has_cross_map', \n",
    "             bins=np.arange(0, clusters_annotated['num_cross_map'].max()),\n",
    "             palette=('k', 'lightgrey'), hue_order=[True, False])\n",
    "\n",
    "# customize the legend\n",
    "legend = ax.get_legend()\n",
    "legend.get_frame().set_visible(False)\n",
    "legend.set_title(\"Removed from QTL analysis?\")\n",
    "# make the plot pretty\n",
    "ax.spines[['top', 'right']].set_visible(False)\n",
    "ax.set_xlabel(\"Number genes in cluster with at least one cross-mappable XX-mer\")\n",
    "ax.set_ylabel(\"Number of clusters\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cluster sharing across tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each pair of genes, are they in a cluster in any other set of tisues?\n",
    "clusters['gene_id'] = clusters['cluster_id'].str.split('_')\n",
    "clusters['partner_gene_id'] = clusters['cluster_id'].str.split('_')f\n",
    "clusters_exploded = clusters.explode('gene_id').explode('partner_gene_id')\n",
    "# remove ones where genes are partnered to themselves \n",
    "clusters_exploded = clusters_exploded[~(clusters_exploded['gene_id'] == clusters_exploded['partner_gene_id'])]\n",
    "# group by pairs of genes\n",
    "gene_partners = clusters_exploded.groupby(['gene_id', 'partner_gene_id']).agg({'cluster_id':'unique', 'tissue_id':'unique'})\n",
    "\n",
    "def get_pretty_tissue_id(tissue_id_list):\n",
    "    return [tissue_id.replace('_', ' ') for tissue_id in tissue_id_list]\n",
    "gene_partners['tissue_id_pretty'] = gene_partners['tissue_id'].apply(get_pretty_tissue_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make an upset plot showing pair sharing\n",
    "upset = up.UpSet(up.from_memberships(gene_partners['tissue_id_pretty']), subset_size='count', min_subset_size=25, show_counts=False, \n",
    "                 sort_by=\"cardinality\", facecolor=\"grey\", totals_plot_elements=3)\n",
    "for i in range(6):\n",
    "    upset.style_subsets(min_degree=i, facecolor=sns.color_palette(\"Greys\")[i])\n",
    "\n",
    "# style the instersection plot\n",
    "plot_result = upset.plot()\n",
    "intersection_plot = plot_result[\"intersections\"]\n",
    "intersection_plot.set_ylabel(\"Shared gene pairs\")\n",
    "intersection_plot.set_xlim(-.8, intersection_plot.get_xlim()[1])\n",
    "intersection_plot.set_ylim(0, intersection_plot.get_ylim()[1]+100)\n",
    "intersection_plot.grid(False)\n",
    "\n",
    "# style the totals plot\n",
    "totals_plot = plot_result[\"totals\"]\n",
    "totals_plot.set_xlim((8000, 0))\n",
    "totals_plot.grid(False)\n",
    "totals_plot.set_xlabel(\"Total gene pairs\")\n",
    "# change totals plot colors to match gtex tissues\n",
    "gtex_tissue_pal_df['tissue_id_pretty'] = gtex_tissue_pal_df['tissue_id'].str.replace('_', ' ')\n",
    "for bar, label in zip(plot_result[\"totals\"].patches, totals_plot.get_yticklabels()):\n",
    "    bar.set_color(gtex_tissue_pal_df[gtex_tissue_pal_df['tissue_id_pretty'] == label.get_text()]['hex'].iloc[0])  # Set the color based on the palette\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
